---
title: "Dennery lab scRNAseq"
author: "Joselynn"
date: "7/14/2020"
output: html_document
---

#add depp1 etc genes back in
#re-order the groups on the x-axis 'Macrophage", "AT2, AT1, Mesenchymal, Endo, other immume

#rerun stats with just findmarkers to  make a specific pairwise comparison to sd7 and o2d7
#in addition to like supplemental table 1, include the actual file name as like a comment
#add some text to the results section with the distributions of the macrophage groups

#push updates to git
#add some AT2 marker gene figures -- do the macrophage seurat clusters have different AT2 markers? could just look at overlaps  

  
```{r "Setup"}
#install.packages('rstatix')
.libPaths(c("/usr/local/lib/R/site-library","/usr/local/lib/R/library"))
library(ggplot2)
library(dplyr)
library(tidyr)
library(Seurat)
library(patchwork)
library(future)
library(rsvd)
library(reshape2)
library(rstatix)
library(biomaRt)
library(stringr)
library(gridExtra)
library(data.table)
library(xlsx)
library(plyr)

set.seed(61)
options(future.globals.maxSize = 4000 * 1024^3)
```
```{r "Data import and merging"}
# These lines don't work if run as part of a markdown code chunk, instead copy and paste them run into the rstudio console and them them there. them via the R console in RStudio (this seems to be an rstudio bug).
# Read in data
setwd('/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq')

AirD7.data <- Read10X(data.dir ="AirD7out/filtered_feature_bc_matrix/")
AirD60.data <- Read10X(data.dir ="AirD60out/filtered_feature_bc_matrix/")
O2D7.data <- Read10X(data.dir ="O2D7out/filtered_feature_bc_matrix/")
O2D60.data <- Read10X(data.dir ="O2D60out/filtered_feature_bc_matrix/")
SD7.data <- Read10X(data.dir ="SD7out/filtered_feature_bc_matrix/")

# Create Seurat objects
#Include features (e.g. genes) detected in at least 3 cells and cells with at least 200 features.
AirD7.data <- CreateSeuratObject(counts = AirD7.data,  project = "AirD7",min.cells = 3, min.features = 200)
O2D7.data <- CreateSeuratObject(counts = O2D7.data,  project = "O2D7",min.cells = 3, min.features = 200)
AirD60.data <- CreateSeuratObject(counts = AirD60.data,  project = "AirD60",min.cells = 3, min.features = 200)
O2D60.data <- CreateSeuratObject(counts = O2D60.data,  project = "O2D60",min.cells = 3, min.features = 200)
SD7.data <- CreateSeuratObject(counts = SD7.data,  project = "SD7",min.cells = 3, min.features = 200)

# Merge the data
all_data <- merge(x = AirD7.data, y = c(O2D7.data, SD7.data, AirD60.data, O2D60.data), add.cell.ids = c("AirD7", "O2D7", "SD7", "AirD60", "O2D60"), project = 'dennery')

# Calculate the percentage of counts that belong to mitochondrial genes
all_data[["percent.mt"]] <- PercentageFeatureSet(all_data, pattern = "^mt-")

#Set levels
Idents(all_data) <- factor(Idents(all_data), levels = c("AirD7","O2D7","SD7", "AirD60", "O2D60"))

saveRDS(object = all_data, file='/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/all_data_seurat.rds')
```
```{r "QC Plots"}
# QC Plots
VlnPlot(all_data, features = "nFeature_RNA")
VlnPlot(all_data, features = "nCount_RNA")
VlnPlot(all_data, features = "percent.mt")
#ggsave(file = '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/figures_and_tables/figures/qc_figures/nFeature_RNA_vlnplot.png', plot = VlnPlot(all_data, features = "nFeature_RNA"))
#ggsave(file = '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/figures_and_tables/figures/qc_figures/nCount_RNA_vlnplot.png', plot = VlnPlot(all_data, features = "nCount_RNA"))
#ggsave(file = '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/figures_and_tables/figures/qc_figures/percent.mt_vlnplot.png', plot = VlnPlot(all_data, features = "percent.mt"))

FeatureScatter(all_data, feature1 = "nCount_RNA", feature2 = "percent.mt")
FeatureScatter(all_data, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

# one QC metric to look at is the number of unique genes detected per cell.
# low-quality/empty droplets will have very few genes or features.
# cell doublets or multiplets will have very high count.
# it looks like there's a cluster of cells in AirD7 and O2D7 where the features and counts are both low -- e.g., there's cells where there's only a few lowly-expressed genes.
# 8000 seems like a reasonable upper bound on the features.
#https://www.bioinformatics.babraham.ac.uk/training/10XRNASeq/seurat_workflow.html
qc.metrics <- dplyr::as_tibble(all_data[[c("nCount_RNA","nFeature_RNA","percent.mt", "orig.ident")]],rownames="Cell.Barcode") 
qc.metrics

#mplot <- qc.metrics %>%
#  arrange(percent.mt) %>%
#  ggplot2::ggplot(aes(nCount_RNA,nFeature_RNA,colour=percent.mt)) + 
#  geom_point() + 
#  scale_color_gradientn(colors=c("black","blue","green2","red","yellow")) +
#  ggtitle("QC metrics") +
#  geom_hline(yintercept = 700) +
#  geom_hline(yintercept = 8000) 

#ident_plot <- qc.metrics %>%
#  arrange(percent.mt) %>%
#  ggplot(aes(nCount_RNA,nFeature_RNA,color=factor(orig.ident))) + 
#  geom_point() + 
#  #scale_color_manual(values=c("#E69F00", "#56B4E9", "#009E73", "#F0E442")) +
#  geom_hline(yintercept = 700) +
#  geom_hline(yintercept = 8000) 

#wrapped_filtering <- wrap_plots(mplot, ident_plot, ncol = 2, guides = "collect") + guides()

#wrapped_filtering
```
```{r "Filtering"}

# Using a 5% mtDNA cutoff... as per https://www.biorxiv.org/content/10.1101/2020.02.20.958793v1.full.pdf
all_data_sub <- subset(all_data, subset = nFeature_RNA > 700 & nFeature_RNA < 8000 & percent.mt < 5)

all_data #20755 features across 16013 samples within 1 assay
all_data_sub #20755 features across 13199 samples within 1 assay 

table(Idents(all_data))
table(Idents(all_data_sub))

```
```{r "Post-filtering  QC plots and normalization"}

FeatureScatter(all_data_sub, feature1 = "nCount_RNA", feature2 = "percent.mt")
FeatureScatter(all_data_sub, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

# SCTransform to normalize, integrate, make some plots
#https://github.com/ChristophH/sctransform/issues/25 for explanation of why I'm not worried about 'iteration limit reached' warning messages.
all_data.list <- SplitObject(all_data_sub, split.by="orig.ident")

#sctransform
all_data.list <- lapply(all_data.list, function(x) {x <- SCTransform(x,verbose=FALSE)})
```
```{r "Integration and clustering"}
#all the integration steps
all_data.features <- SelectIntegrationFeatures(object.list=all_data.list,nfeatures=3000)
all_data.list <- PrepSCTIntegration(object.list=all_data.list, anchor.features=all_data.features, verbose=FALSE)
all_data.int.anchors <- FindIntegrationAnchors(all_data.list, normalization.method="SCT", anchor.features=all_data.features, verbose=FALSE)
all_data.combined <- IntegrateData(anchorset=all_data.int.anchors, normalization.method="SCT", verbose=FALSE)

DefaultAssay(all_data.combined) <- "integrated"

all_data.combined <- RunPCA(all_data.combined, features = VariableFeatures(object = all_data.combined))
all_data.combined <- FindNeighbors(all_data.combined, dims = 1:50)
all_data.combined <- FindClusters(all_data.combined, resolution = 2)
all_data.combined <- RunUMAP(all_data.combined, dims=1:50)
all_data.combined <- RunTSNE(all_data.combined, dims=1:50, nthreads=4, max_iter=2000, seed.use=35) 
#DimPlot(all_data.combined, reduction="tsne")
#all_data.combined <- RunTSNE(all_data.combined, dims=1:50, tsne.method = "FIt-SNE", nthreads=4, max_iter=2000, seed.use=35)  #Could not find the following programs: fast_tsne

```
```{r "Import Mouse cell atlas data"}
# MCA
#From https://satijalab.org/seurat/v3.1/mca.html vignette...

setwd('/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq')

mca.matrix <- readRDS(file = "/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/MCA/MCA_merged_mat.rds")
mca.metadata <- read.csv("/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/MCA/MCA_All-batch-removed-assignments.csv")
mca.annotations <- read.csv("/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/MCA/MCA_CellAssignments.csv",row.names = 1)

# just interested in lung tissue
mca.annotations.lung <- mca.annotations %>% dplyr::filter(Tissue == 'Lung')
mca.metadata.lung <- mca.metadata %>% dplyr::filter(Tissue == 'Lung')

rownames(mca.annotations.lung) <- mca.annotations.lung$Cell.name
rownames(mca.metadata.lung) <- mca.metadata.lung$Cell.name

mca.metadata.annotations <- dplyr::inner_join(mca.metadata.lung, mca.annotations.lung, by = "Cell.name")
mca.metadata.annotations <- mca.metadata.annotations %>% dplyr::select(Cell.name, ClusterID.x, Tissue.x, Batch.x, Cell.Barcode.x, Annotation)
colnames(mca.metadata.annotations) <- c('Cell.name', 'ClusterID', 'Tissue', 'Batch', 'Cell.Barcode', 'Annotation')
rownames(mca.metadata.annotations) <- mca.metadata.annotations$Cell.name

mca.metadata.annotations <- mca.metadata.annotations %>% dplyr::select(-Cell.name)

mca.matrix.lung <- mca.matrix[,grepl("^Lung_", colnames(mca.matrix))]
mca <- CreateSeuratObject(counts = mca.matrix.lung, meta.data = mca.metadata.annotations, project = "MouseCellAtlas")

#saveRDS(mca, file = "mca_sept_2021.rds")
#saveRDS(all_data.combined, file = "all_data.combined_sept_2021.rds")

```
```{r "Normalize mca, transfer anchors"}
# Only keep annotated cells
mca <- SCTransform(mca)
# Leaves us with 242k cells
```

```{r}
all_data.list$mca <- mca
atlas.features <- SelectIntegrationFeatures(all_data.list) #might need to tweak this 
atlas.anchors <- FindTransferAnchors(reference=mca, query=all_data.combined, normalization.method="SCT", npcs=82, dims=1:50, features=atlas.features, verbose=FALSE)

atlas.predictions <- TransferData(anchorset=atlas.anchors, refdata=mca$Annotation, dims=1:50)
all_data.combined <- AddMetaData(all_data.combined, metadata=atlas.predictions)
all_data.combined$atlas_predicted_clusters <- sapply(sapply(all_data.combined$predicted.id, function(x) strsplit(x, " - ")), `[`, 1)

all_data.combined@meta.data$atlas_predicted_clusters_short <- str_replace(all_data.combined@meta.data$atlas_predicted_clusters, pattern = "\\(Lung\\)", replacement = "")


#saveRDS(mca, file = "mca_sept_2021.rds")
#saveRDS(all_data.combined, file = "all_data.combined_sept_2021.rds")

all_data.combined@meta.data$seurat_mca <- paste0(all_data.combined@meta.data$seurat_clusters, '_', all_data.combined@meta.data$atlas_predicted_clusters_short)
table(subset(all_data.combined, subset = orig.ident == 'SD7')@meta.data$seurat_mca)


table(subset(all_data.combined, subset = orig.ident == 'SD7')@meta.data$seurat_clusters, subset(all_data.combined, subset = orig.ident == 'SD7')@meta.data$atlas_predicted_clusters_short)

table(all_data.combined@meta.data$seurat_clusters, all_data.combined@meta.data$atlas_predicted_clusters_short)

table(subset(all_data.combined, subset = orig.ident == 'SD7')@meta.data$seurat_clusters)
table(subset(all_data.combined, subset = orig.ident == 'SD7')@meta.data$atlas_predicted_clusters_short)

table(all_data.combined@meta.data$orig.ident)

table(subset(all_data.combined, subset = orig.ident == 'O2D7')@meta.data$seurat_clusters)
table(subset(all_data.combined, subset = orig.ident == 'O2D7')@meta.data$atlas_predicted_clusters_short)




```
```{r "Fix 'Clara' to 'Club'"}

all_data.combined$atlas_predicted_clusters <- str_replace(all_data.combined$atlas_predicted_clusters, pattern = 'Clara', replacement = 'Club')
all_data.combined$predicted.id <- str_replace(all_data.combined$predicted.id, pattern = 'Clara', replacement = 'Club')
all_data.combined$atlas_predicted_clusters_short <- str_replace(all_data.combined$atlas_predicted_clusters_short, pattern = 'Clara', replacement = 'Club')
colnames(all_data.combined@meta.data) <- str_replace(colnames(all_data.combined@meta.data), pattern = 'Clara', replacement = 'Club')

all_data.combined$orig.ident <- factor( x = all_data.combined$orig.ident, levels = c("AirD7", "O2D7", "SD7",  "AirD60", "O2D60"))

#saveRDS(mca, file = "mca_sept_2021.rds")
#saveRDS(all_data.combined, file = "all_data.combined_sept_2021.rds")


```
```{r "Add another level of cell type annotations as per Hongwei"}
#assign a new column called atlas_predicted_clusters_grouped
all_data.combined$atlas_predicted_clusters_grouped <- all_data.combined$atlas_predicted_clusters_short

#find and replace to get new levels in
all_data.combined$atlas_predicted_clusters_grouped <- str_replace_all(string = all_data.combined$atlas_predicted_clusters_grouped,
            c("Alveolar macrophage_Ear2 high" = "Macrophages",
            "Alveolar macrophage_Pclaf high" = "Macrophages",
            "Interstitial macrophage" = "Macrophages",
            "Basophil" = "Other immune cells",
            "T Cell_Cd8b1 high" = "Other immune cells",
            "B Cell" = "Other immune cells",
            "Nuocyte" = "Other immune cells",
            "Conventional dendritic cell_Gngt2 high" = "Other immune cells",
            "Eosinophil granulocyte" = "Other immune cells",
            "Plasmacytoid dendritic cell" = "Other immune cells",
            "Dendritic cell_Naaa high" = "Other immune cells",
            "Dividing T cells" = "Other immune cells",
            "NK Cell" = "Other immune cells",
            "Conventional dendritic cell_H2-M2 high" = "Other immune cells",
            "Neutrophil granulocyte" = "Other immune cells",
            "Conventional dendritic cell_Mgl2 high" = "Other immune cells",
            "Conventional dendritic cell_Tubb5 high" = "Other immune cells",
            "Dividing dendritic cells" = "Other immune cells",
            "Ig−producing B cell" = "Other immune cells",
            "AT1 Cell" = "AT1 cells",
            "AT2 Cell" = "AT2 cells",
            "Ciliated cell" = "Other epithelial cells",
            "Dividing cells" = "Other epithelial cells",
            "Club Cell" = "Other epithelial cells",
            "Stromal cell_Acta2 high" = "Mesenchymal cells",
            "Stromal cell_Inmt high" = "Mesenchymal cells",
            "Stromal cell_Dcn high" = "Mesenchymal cells",
            "Endothelial cell_Tmem100 high" = "Endothelial cells",
            "Endothelial cell_Kdr high" = "Endothelial cells",
            "Endothelial cells_Vwf high" = "Endothelial cells"))

all_data.combined$seurat_mca <- paste0(all_data.combined$seurat_clusters, '-', all_data.combined$atlas_predicted_clusters_grouped)


all_data.combined$atlas_predicted_clusters_grouped <- factor(all_data.combined$atlas_predicted_clusters_grouped, levels = c("Macrophages", "AT2 cells", "AT1 cells", "Mesenchymal cells", "Endothelial cells", "Other epithelial cells", "Other immune cells"))
```
```{r "make sd7 distributions table with fixed cell type ids"}
sd7_distributions <- data.frame((table(subset(all_data.combined, subset = orig.ident == 'SD7')@meta.data$seurat_clusters, subset(all_data.combined, subset = orig.ident == 'SD7')@meta.data$atlas_predicted_clusters_short))) %>% pivot_wider(names_from = 'Var1', values_from = 'Freq')
#write.xlsx(sd7_distributions, file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/sd7_distributions.xlsx')

sd7_distributions_grouped <- data.frame((table(subset(all_data.combined, subset = orig.ident == 'SD7')@meta.data$seurat_clusters, subset(all_data.combined, subset = orig.ident == 'SD7')@meta.data$atlas_predicted_clusters_grouped))) %>% pivot_wider(names_from = 'Var1', values_from = 'Freq')
#write.xlsx(sd7_distributions_grouped, file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/sd7_distributions_grouped.xlsx')

```
```{r "Look at TSNE plots and compare mca vs seurat clusters"}

p1 <- DimPlot(all_data.combined, reduction="tsne", group.by = "atlas_predicted_clusters") 
p1 <- LabelClusters(p1, id="atlas_predicted_clusters", repel=TRUE)
#ggsave(plot = p1, file = '2a1_nolabel.png', width = 10, height = 10, units = 'in')
#p1

p2 <- DimPlot(all_data.combined, reduction="tsne", group.by = "seurat_clusters") #+ NoLegend()
p2 <- LabelClusters(p2, id="seurat_clusters", repel=TRUE)
#ggsave(plot = p2, file = '2a2_nolabel.png', width = 10, height = 10, units = 'in')
#p2


#wrap_plots(list(p1,p2), ncol = 2)

p3 <- DimPlot(all_data.combined, reduction="tsne", group.by = "atlas_predicted_clusters_short", split.by = "orig.ident") 
p3 <- LabelClusters(p3, id="atlas_predicted_clusters_short", repel=TRUE)
#p3

p4 <- DimPlot(all_data.combined, reduction="tsne", group.by = "seurat_clusters", split.by = "orig.ident") 
p4 <- LabelClusters(p4, id="seurat_clusters", repel=TRUE)
#p4
```
```{r "Get more info per each gene"}

# get a list of all the genes in the Seurat object.
all_genes <- data.frame(rownames(all_data.combined))
all_genes$mgi_symbol <- all_genes$rownames.all_data.combined.
all_genes$rownames.all_data.combined. <- NULL

mouse_biomart <- useMart("ensembl", dataset = "mmusculus_gene_ensembl", host="apr2022.archive.ensembl.org")

mouse_biomart_data <- getBM(mart = mouse_biomart, attributes = c("mgi_description", "mgi_symbol", "reactome"))
all_genes_bm <- dplyr::right_join(mouse_biomart_data, all_genes)


all_genes_bm <- ddply(all_genes_bm, .(mgi_description, mgi_symbol), summarise, reactome_ids = paste0(reactome, collapse = ", "))


#saveRDS(all_genes_bm, "all_genes_bm.rds")
#all_genes_bm <- readRDS("/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/all_genes_bm.RDS")

```
```{r "look at at1 at2 etc marker genes in macrophages"}
#bin the cell types
#Endothelial all, #macrophage all, 
#"Alveolar macrophage_Ear2 high", "Interstitial macrophage", "Alveolar macrophage_Pclaf high"

#a)	type II cell markers (sftpc, sftpb, and abca3), type I cell biomarkers (Hpox, AQP5, andT1α), mesenchymal #cells (Pdgfra) or SOX9 positive cells.
#b)	genes involved in macrophage phagocytosis (Timd4, Mer, Mfge8, AnxA1, Nr1h3, Nr1h2 , and Nr1c3). See this #paper: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5413334/


Idents(all_data.combined) <- 'seurat_clusters'

```
```{r "Try making some more compelling featureplots for the genes Hongwei suggested}

#all_genes_bm <- readRDS('all_genes_bm.RDS')

#Please see below the gene list which can be used for identification of senescence. 
#Gene list of senescence:
#Cdkn1a (encoding p21 protein), cdkn2a (encoding p14, p16, and p19 proteins), cdkn2b (encoding p15 protein), p53, Rb, e2f.
#We may also want to see genes (ki67, pcna) of proliferation, which is usually decreased in senescent cells.
#Some genes including sirt1, sirt3, sirt6, lmnb1, egr1, ezh2 are also associated with senescence
#Genes of senescence-associated secretory phenotype (SASP) include IL-1, IL-6, IL-10, PAI1, TNF-α, CXCL1/KC, CXCL2, CCL3, CCL2, and CCL2/MCP-1, MMP3, 
cdkn <- grep('cdkn', all_genes_bm$mgi_symbol, ignore.case = T, value = TRUE)
p53 <- grep('p53', all_genes_bm$mgi_symbol, ignore.case = T, value = TRUE)
#didn't find 'Rb'
e2f <- grep('e2f', all_genes_bm$mgi_symbol, ignore.case = T, value = TRUE)
ki67 <- grep('ki67', all_genes_bm$mgi_symbol, ignore.case = T, value = TRUE)
pcna <- grep('pcna', all_genes_bm$mgi_symbol, ignore.case = T, value = TRUE)
#didn't find 'Sirt'
lmnb1 <- grep('lmnb1', all_genes_bm$mgi_symbol, ignore.case = T, value = TRUE)
egr1 <- grep('egr1', all_genes_bm$mgi_symbol, ignore.case = T, value = TRUE)
ezh2 <- grep('ezh2', all_genes_bm$mgi_symbol, ignore.case = T, value = TRUE)
il6 <-'Il6'
il1 <- c('Il1a','Il1b')  
il10 <- c('Il10rb', 'Il10ra')
#serpine is synonym for pai1 http://www.informatics.jax.org/marker/MGI:97608
serpine1 <- grep('Serpine1', all_genes_bm$mgi_symbol, ignore.case = T, value = TRUE)
tnfs <- grep('tnf', all_genes_bm$mgi_symbol, ignore.case = T, value = TRUE)
cxc <- grep('cxc', all_genes_bm$mgi_symbol, ignore.case = T, value = TRUE)
ccl <- grep('ccl', all_genes_bm$mgi_symbol, ignore.case = T, value = TRUE)
mmp <- grep('mmp', all_genes_bm$mgi_symbol, ignore.case = T, value = TRUE)

DefaultAssay(all_data.combined) <- 'SCT'
Idents(all_data.combined) <- 'seurat_clusters'
                        
O2D7_seurat <- subset(all_data.combined, orig.ident == 'O2D7')
AirD7_seurat <- subset(all_data.combined, orig.ident == 'AirD7')
SD7_seurat <- subset(all_data.combined, orig.ident == 'SD7')
O2D60_seurat <- subset(all_data.combined, orig.ident == 'O2D60')
AirD60_seurat <- subset(all_data.combined, orig.ident == 'AirD60')


Idents(O2D7_seurat) <- 'seurat_clusters'
Idents(AirD7_seurat) <- 'seurat_clusters'
Idents(SD7_seurat ) <- 'seurat_clusters'
Idents(O2D60_seurat) <- 'seurat_clusters'
Idents(AirD60_seurat) <- 'seurat_clusters'
```
```{r "actually make plots"}

featureplot_grid <- function(feats, pwidth, pheight, plotfile){
  p_aird7 <- FeaturePlot(AirD7_seurat, 
                            label = TRUE, 
                            label.size = 6, 
                            repel = F, 
                            sort = TRUE, 
                            features = feats, 
                            ncol = length(feats), 
                            reduction = 'tsne') &
  ggplot2::scale_colour_gradientn(limits= c(0,8), 
                         colors = c('bisque', 'darkred')) &
  labs(subtitle = 'Air D7')

p_o2d7 <- FeaturePlot(O2D7_seurat, 
                            label = TRUE, 
                            label.size = 6, 
                            repel = F, 
                            sort = TRUE, 
                            features = feats, 
                            ncol = length(feats), 
                            reduction = 'tsne') &
  ggplot2::scale_colour_gradientn(limits= c(0,8), 
                         colors = c('bisque', 'darkred')) &
  labs(subtitle = 'O2 D7')

p_sd7 <- FeaturePlot(SD7_seurat, 
                            label = TRUE, 
                            label.size = 6, 
                            repel = F, 
                            sort = TRUE, 
                            features = feats, 
                            ncol = length(feats), 
                            reduction = 'tsne') &
  ggplot2::scale_colour_gradientn(limits= c(0,8), 
                         colors = c('bisque', 'darkred')) &
  labs(subtitle = 'S D7')

p_aird60 <- FeaturePlot(AirD60_seurat, 
                            label = TRUE, 
                            label.size = 6, 
                            repel = F, 
                            sort = TRUE, 
                            features = feats, 
                            ncol = length(feats), 
                            reduction = 'tsne') &
  ggplot2::scale_colour_gradientn(limits= c(0,8), 
                         colors = c('bisque', 'darkred')) &
  labs(subtitle = 'Air D60')

p_o2d60 <- FeaturePlot(O2D60_seurat, 
                            label = TRUE, 
                            label.size = 6, 
                            repel = F, 
                            sort = TRUE, 
                            features = feats, 
                            ncol = length(feats), 
                            reduction = 'tsne') &
  ggplot2::scale_colour_gradientn(limits= c(0,8), 
                         colors = c('bisque', 'darkred')) &
  labs(subtitle = 'O2 D60')


plots <- p_aird7 / p_o2d7 / p_sd7 / p_aird60 / p_o2d60

#ggsave(plotfile, device = 'png', width = pwidth, height = pheight)
  
return(plots)
}

featureplot_grid(cdkn, pwidth = 30, pheight = 30, 'p_cdkn_fp.png')
featureplot_grid(p53, pwidth = 6, pheight = 30, 'p_p53_fp.png')
featureplot_grid(e2f, pwidth = 15, pheight = 30, 'p_e2f_fp.png')
featureplot_grid(ki67, pwidth = 6, pheight = 30, 'p_ki67_fp.png')
featureplot_grid(pcna, pwidth = 6, pheight = 30, 'p_pcna_fp.png')
featureplot_grid(lmnb1, pwidth = 6, pheight = 30, 'p_lmnb1_fp.png')
featureplot_grid(egr1, pwidth = 6, pheight = 30, 'p_egr1_fp.png')
featureplot_grid(ezh2, pwidth = 6, pheight = 30, 'p_ezh2_fp.png')
featureplot_grid(il6, pwidth = 6, pheight = 30, 'p_il6_fp.png')
featureplot_grid(il1, pwidth = 15, pheight = 30, 'p_il1_fp.png')
featureplot_grid(il10, pwidth = 15, pheight = 30, 'p_il10_fp.png')
featureplot_grid(serpine1, pwidth = 6, pheight = 30, 'p_serpine1_fp.png')
featureplot_grid(tnfs, pwidth = 40, pheight = 30, 'p_tnfs_fp.png')
featureplot_grid(cxc, pwidth = 40, pheight = 30, 'p_cxc_fp.png')
featureplot_grid(mmp, pwidth = 40, pheight = 30, 'p_mmp_fp.png')
```
```{r "Make a dotplot of Hongwei's genes of interest"}


features_list <- c(cdkn, p53, e2f, ki67, pcna, lmnb1, egr1, ezh2, il6, il1, il10, serpine1, tnfs, cxc, mmp, 'Sin3a', 'Plaur', 'B2m', 'Icam1', 'Dpp4', 'Depp1')

#features_list <- c(cdkn, p53, e2f, ki67, pcna, lmnb1, egr1, ezh2, il6, il1, il10, serpine1, tnfs, cxc, mmp)

features_list <- unique(features_list)

features_list <- factor(features_list, levels = c("Mmp12","Mmp8","Mmp11","Mmp23","Mmp14","Mmp19","Mmp2","Mmp25","Cxcr2","Cxcl10","Cxcl2","Cxcl1","Cxcl15","Cxcl14","Cxcr4","Tnfrsf1b","C1qtnf6","Tnfrsf18","Tnfrsf4","Tnfaip3","Tnfrsf13b","Tnfsf14","Tnfsf9","Tnfaip8l2","Tnfrsf12a","Tnfaip2","Serpine1","Il10ra","Il10rb","Il1b","Il1a","Il6","Ezh2","Egr1","Lmnb1","Pcna","Mki67","E2f8","E2f7","E2f1","Trp53i11","Cdkn1c","Cdkn2d","Cdkn1a","Cdkn3","Cdkn2b","Cdkn2c", "Sin3a", "Plaur", "B2m", "Icam1", "Dpp4", "Depp1"))

#table(subset(all_data.combined, subset = orig.ident == 'SD7')@meta.data$atlas_predicted_clusters_grouped)
#table(subset(all_data.combined, subset = orig.ident == 'SD7')@meta.data$seurat_clusters)
#table(subset(all_data.combined, subset = orig.ident == 'SD7')@meta.data$seurat_mca)

sd7_freq_table <- data.frame(table(
      subset(all_data.combined, subset = orig.ident == 'SD7')$seurat_mca)) %>% arrange(desc(Freq))

sd7_freq_table

table(subset(all_data.combined, subset = orig.ident == 'SD7')@meta.data$atlas_predicted_clusters_grouped)

#clusters 17, 21, 13, 23, 15, 16, 25, and 1 are macrophage groups.

#do the macrophages express senesence marker genes? e.g., sasp
#do the macrophages express surfactant or other at2-ish genes


#add sin3a, sox9 and pgf genes
#are these genes makers for different cell types?
```
```{r "make plots"}
Idents(all_data.combined) <- 'orig.ident'


all_dotplot <-  DotPlot(all_data.combined, assay = 'SCT',features = features_list) & ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) & coord_flip() & theme(axis.text=element_text(size=20))

#ggsave('/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/types_dotplot.png', plot = types_dotplot, width = 12, height = 14)

Idents(all_data.combined) <- 'atlas_predicted_clusters_grouped'

types_dotplot <-  DotPlot(all_data.combined, assay = 'SCT',features = features_list ) & ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) & coord_flip()& theme(axis.text=element_text(size=20))

types_dotplot_sub <-  DotPlot(subset(all_data.combined, subset = orig.ident == 'SD7'), assay = 'SCT',features = features_list ) & ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) & coord_flip() & theme(axis.text=element_text(size=20), legend.text=element_text(size=20), axis.title = element_blank())


types_dotplot_flip <-  DotPlot(all_data.combined, assay = 'SCT',features = features_list) & ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))& theme(axis.text=element_text(size=20))

types_dotplot_sub_flip <-  DotPlot(subset(all_data.combined, subset = orig.ident == 'SD7'), assay = 'SCT',features = features_list ) & ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))& theme(axis.text=element_text(size=20))

Idents(all_data.combined) <- 'orig.ident'


types_dotplot_ident <-  DotPlot(all_data.combined, assay = 'SCT',features = features_list) & ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) & coord_flip()& theme(axis.text=element_text(size=20))


#ggsave('/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/types_dotplot.png', plot = types_dotplot, width = 14, height = 14)

#ggsave('/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/types_dotplot_sd7.png', plot = types_dotplot_sub, width = 14, height = 14)

#ggsave('/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/all_dotplot.png', plot = all_dotplot, width = 7, height = 12)

#ggsave('/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/types_dotplot_sub.png', plot = types_dotplot_sub, width = 14, height = 18)


p_cdkn_dot <- DotPlot(all_data.combined, assay = 'SCT',features = unique(cdkn))
p_p53_dot <- DotPlot(all_data.combined, assay = 'SCT',features = unique(p53))
p_e2f_dot <- DotPlot(all_data.combined, assay = 'SCT',features = unique(e2f))
p_ki67_dot <- DotPlot(all_data.combined, assay = 'SCT',features = unique(ki67))
p_pcna_dot <- DotPlot(all_data.combined, assay = 'SCT',features = unique(pcna))
p_lmnb1_dot <- DotPlot(all_data.combined, assay = 'SCT',features = unique(lmnb1))
p_egr1_dot <- DotPlot(all_data.combined, assay = 'SCT',features = unique(egr1))
p_ezh2_dot <- DotPlot(all_data.combined, assay = 'SCT',features = unique(ezh2))
p_il6_dot <- DotPlot(all_data.combined, assay = 'SCT',features = unique(il6))
p_il1_dot <- DotPlot(all_data.combined, assay = 'SCT',features = unique(il1))
p_il10_dot <- DotPlot(all_data.combined, assay = 'SCT',features = unique(il10))
p_serpine1_dot <- DotPlot(all_data.combined, assay = 'SCT',features = unique(serpine1))

p_tnfs_dot <- DotPlot(all_data.combined, assay = 'SCT',features = unique(tnfs))
p_cxc_dot <- DotPlot(all_data.combined, assay = 'SCT',features = unique(cxc))
p_mmp_dot <- DotPlot(all_data.combined, assay = 'SCT',features = unique(mmp))

```
```{r "1C metab"}
onecarbon_genes <- c('PHGDH','PSAT1','PSPH','DHFR','SHMT','MTHFR','MTHFD','MS','MAT','SAHH','CBS')

onecarbon_genes_more <- all_genes_bm %>% dplyr::filter(grepl('R-MMU-196757', reactome_ids))

onecarbon_genes_fixed <- lapply(seq_along(onecarbon_genes), function(i) grep(onecarbon_genes[i], x = rownames(all_data.combined), ignore.case = T, value = T)) 

onecarbon_genes_fixed <- c("Phgdh", "Psat1", "Psph", "Dhfr", "Shmt2", "Shmt1", "Mthfr", "Mthfd2l", "Mthfd2","Mthfd1l", "Mthfd1", "Mtr", "Ahcyl1", "Ahcyl2", "Cbs", "Folr2")



one_carbon_genes_sd7_mca <- DotPlot(subset(all_data.combined, subset = orig.ident == 'SD7'), features = onecarbon_genes_fixed, group.by = 'atlas_predicted_clusters_short', assay = 'SCT') & theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))


one_carbon_genes_idents <-  DotPlot(all_data.combined, assay = 'SCT',features =  onecarbon_genes_fixed) & theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 



#ggsave('/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/one_carbon_genes_sd7_only_mca.png', plot = one_carbon_genes_sd7_mca, width = 8, height = 8)

#ggsave('/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/one_carbon_genes_idents.png', plot = one_carbon_genes_idents , width = 8, height = 8)



```
```{r "specific comparison of sd7 and o2d7 plus all condition markers w findallmarkers"}
all_data.combined <- PrepSCTFindMarkers(all_data.combined, assay = 'SCT', verbose = TRUE)

#condition markers for all data
Idents(all_data.combined) <- 'orig.ident'

all_condition_markers <- FindAllMarkers(all_data.combined, min.pct = 0.25, logfc.threshold = 0.25, assay = 'SCT')

#sd7 specific markers
SD7_markers <- all_condition_markers %>% dplyr::filter(cluster == 'SD7')

#find markers comparing SD7 and O2D7
sd7_vs_o2d7 <- FindMarkers(all_data.combined, ident.1 = 'SD7', ident.2 = 'O2D7' , assay = 'SCT')
 
#find markers comparing SD7 and airD7
sd7_vs_aird7 <- FindMarkers(all_data.combined, ident.1 = 'SD7', ident.2 = 'AirD7' , min.pct = 0.25, logfc.threshold = 0.25, assay = 'SCT')

#find markers comparing o27 and airD7
o2d_vs_aird7 <- FindMarkers(all_data.combined, ident.1 = 'O2D7', ident.2 = 'AirD7' , min.pct = 0.25, logfc.threshold = 0.25, assay = 'SCT')


#look at cluster markers for all data
Idents(all_data.combined) <- 'seurat_clusters'
all_cluster_markers <- FindAllMarkers(all_data.combined, min.pct = 0.25, logfc.threshold = 0.25, assay = 'SCT')

#cluster markers for only SD7 data.
#don't recorrect umi since they are from the same sct object

SD7_cluster_markers <- FindAllMarkers(SD7_seurat, min.pct = 0.25, logfc.threshold = 0.25, assay = 'SCT',recorrect_umi = FALSE)

#add gene ids to sd7_vs_o2d7 and sd7_vs_aird7
sd7_vs_o2d7$gene <- rownames(sd7_vs_o2d7)
sd7_vs_aird7$gene <- rownames(sd7_vs_aird7)
o2d_vs_aird7$gene <- rownames(o2d_vs_aird7)
```
```{r "get some more genes of interest"}
AT2_markers <- unique(c("Abhd5", "Acaa1a", "Adgrf5", "Aig1", "Arhgap44", "Atf3", "Atp13a3", "Bmp4", "Bnip3", "Ccz1", "Cdh16", "Cds1", "Cftr", "Clu", "Cndp2", "Col6a1", "Cracr2a", "Cyp51", "Dhcr24", "Dpp9", "Echs1", "Eif3b", "Epb41l5", "Erbb3", "Fam126b", "Fdft1", "Fgf1", "Foxp2", "Gata6", "Gsta4", "Hadh", "Hdc", "Higd1a", "Hmgcr", "Igsf3", "Il18r1", "Irs2", "Itih4", "Junos", "Kcne2", "Kcnj13", "Kcnj2", "Klf9", "LOC115487124", "Lratd2", "Maged1", "Malat1", "Mob3b", "Mrps9", "Myh7", "Nedd4l", "P4ha1", "Paqr8", "Pard3", "Pcnt", "Pcyt2", "Pls3", "Plxnb1", "Pof1b", "Ppp1r16a", "Rac3", "Rbms3", "Rn7sk", "Serpinb9", "Sgpp2", "Sh3d19", "Slc16a7", "Slc25a39", "Sowahc", "Spint1", "Spr", "Spred1", "Sqle", "Srebf2", "Steap2", "Syt7", "Tbc1d30", "Tcea3", "Tkt", "Tmc4", "Tmem205", "Tmem8", "Tmprss4", "Trim2", "Trp53inp2", "Trpm6", "Ttc38", "Txnrd3", "Usp33", "Vsig2","Rtkn2", "Spock2", "Col4a3", "2200002D01Rik", "Sema3e", "Lamc2", "Scnn1g", "Marco", "Itgb6", "Lama3", "Ccl6", "Igfbp7", "Fth1", "Fbln5", "Krt19", "Sec14l3", "Cd9", "Ano1", "Lpl", "Cryab", "F3", "Tmsb4x", "Vegfa", "Clic5", "S100a11", "Anxa2", "Top2a", "Myl6", "Col4a1", "Actb", "Hs2st1", "Agrn", "Ear2", "Plet1", "Fam174b", "Atp6v0d2", "Fzd2", "Nnat", "Pclaf", "Hspg2", "Rpl36al", "Mest", "Rps27a", "Ctsd", "Cstb", "Akr1a1", "Col4a2", "Calm1", "Fads3", "Plin2", "Tspan15", "Ndufa4", "Ndufb1-ps", "Rpl19", "Aqp1", "Rpl26", "Rpl11", "Ywhae", "Cox4i1", "Sem1", "Anxa5", "Fam189a2", "Uqcrb", "Eif1", "Rps20", "Rpl35", "Calm2", "Ftl1", "Ndst1", "Atp5l", "Cox8a", "Col4a5", "H3f3b", "Serpinh1", "Btf3", "Serf2", "Rpl34", "Tmsb10", "Itgb1", "Cox5b", "Tagln2", "Rpl13", "Rpl14", "Rpl22l1", "Rpl36", "Hsp90ab1", "Trp53bp2", "Myl12b", "Atp5j", "Rab11fip1", "Ctsb", "Lamp1", "Ptov1", "Rps4x", "Rpl22", "Atp5pb", "Ctsz", "Rps13", "Ehd2", "Rpl35a", "Hspa8", "Rps24", "S100a1", "Rps3a1", "Lrpap1", "Drap1", "Rhoa", "Pdia3", "Ndufa13", "Atp5md", "Cd81", "Cox5a", "F11r", "Rpl7", "Cttnbp2nl", "Mrpl14", "Eif2s2", "Ybx1", "1810037I17Rik", "Arf1", "Rpl37a", "Rps23", "Tppp3", "Tmem176b", "Atp5mpl", "Rpl18", "Lamc1", "Rpl36a", "Atp5b", "Npm1", "Ahnak", "Tomm6", "Rpl27", "Rpl9", "Anxa1", "Rpl23", "Rpl8", "Mmp14", "Car2", "Car2", "Rps3", "Canx", "Rps15a", "Xist", "Paip1", "Atp5g1", "Rpl27a", "Tomm7", "Uba52", "Rps27l", "Myof", "H3f3a", "Hspa5", "Slc25a4", "Pmp22", "Tubb4b", "Tmed10", "Hmgb2", "Gng5", "Smdt1", "Atp5c1", "Laptm4a", "Spcs2", "Edf1", "Ctdspl", "Eef1a1", "Eif4a1", "Mpc2", "Crip1", "Slc25a3", "Rpl37", "Rps8", "Ubl5", "Rpl38", "Gnb1", "Rpl28", "Ndufb9", "Gnb2", "LOC102636299", "Cyba", "Qk", "Eif3f", "Rpl30", "Dad1", "Eef2", "Esd", "Pcna", "Pfn1", "Ndufa5", "Hsp90b1", "Actn4", "Uqcrq", "Tuba1a", "Ndufa3", "Cnn2", "Tmem256", "Galnt18", "Ndufb7", "Uqcr10", "Snrpe", "Cux1", "Ndufb8", "Sptbn1", "Atp5g3", "Arf5", "Ndufa8", "Hnrnpa3", "Rpl37rt", "Selenow", "Ndufa1", "Psap", "Prdx5", "Rps9", "Ost4", "Atp5o", "Atp6v0d1", "Selenoh", "Clta", "Reep5", "Mrfap1", "Ndufb5", "Pon2", "Hypk", "Pomp", "Sox4", "Mrps21", "Ssr4", "Dctn3", "Tpm3", "Rpl17", "Fdx1", "Rps15a-ps6", "Aldoa", "Hsbp1", "Ndufb2", "Cox7b", "Fermt2", "Serbp1", "Acaa2", "Psmb2", "Nedd8", "Morf4l1", "Uqcrh", "Ndufc1", "Cdk4", "Hnrnpk", "Abcg1", "Ppp2cb", "Cycs", "Vdac2", "Ndufa2", "Tjp1", "Fis1", "Ran", "Cavin1", "Cdc42", "Aprt", "Eif3h", "Romo1", "Atp6v1f", "Ywhah", "Snrpd2", "Cfl1", "Tma7", "Slc39a1", "Nme1", "Cbx3", "Rbx1", "Micos10", "Ndufb11", "Rtraf", "Krtcap2", "Ubc", "Ifi27", "Ostc", "Alyref", "Mlf2", "Arhgap5", "Rab5if", "Dynlrb1", "Svil", "Lamtor2", "Tbca", "Snrpg", "Ctnna1", "Psma7", "Dpysl2", "Psmb1", "Fkbp1a", "1810058I24Rik", "Itm2c", "Hdgf", "Plp2", "Hadha", "Ndufs6", "Hmgn2", "Capns1", "D8Ertd738e", "Tpm4", "Gm9385", "Swi5", "Pdcd5", "Pebp1", "Anp32a", "Mpc1", "Tomm20", "Atp6v0e", "Erh", "Itm2b", "Mif", "Arpp19", "Psma2", "Cox17", "Psmb3", "Tmed2", "Skp1a", "Ywhaq", "Vcp", "Spcs1", "Kdelr2", "Psmd8", "Ube2d3", "Hnrnpa0", "Ndufa11", "Pkm", "Dync1i2", "Emd", "Psmb5", "Bzw1", "Tmem258", "Hnrnpm", "Rhob", "Hnrnpd", "Selenok", "Alas2", "Alas2", "Igfbp5", "Ube2c", "Gpc3", "Bpgm", "Bpgm", "Bpgm", "Bnip3l"))

```
```{r "Look at the sd7 condition markers and filter to pull out at2, sasp, 1c"}


#look at condition markers and pull out markers for SD7, then compre to AT2 markers, SASP genes, and 1C genes.
SD7_markers_sasp <- dplyr::filter(SD7_markers, gene %in% features_list )
SD7_markers_onecarbon <- dplyr::filter(SD7_markers, gene %in% onecarbon_genes_fixed)
SD7_markers_at2 <- dplyr::filter(SD7_markers, gene %in% AT2_markers)

#write.xlsx(SD7_markers, file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/SD7_markers.xlsx', row.names = F)
#write.xlsx(SD7_markers_sasp, file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/SD7_markers_sasp.xlsx', row.names = F)
#write.xlsx(SD7_markers_onecarbon , file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/SD7_markers_onecabon.xlsx', row.names = F)
#write.xlsx(SD7_markers_at2 , file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/SD7_markers_at2.xlsx', row.names = F)

#what about filtering from the sd7 vs o2d7 contrasts and other pairwise contrasts:

sd7_vs_o2d7_markers_sasp <- dplyr::filter(sd7_vs_o2d7, gene %in% features_list )
sd7_vs_o2d7_markers_onecarbon <- dplyr::filter(sd7_vs_o2d7, gene %in% onecarbon_genes_fixed)
sd7_vs_o2d7_markers_at2 <- dplyr::filter(sd7_vs_o2d7, gene %in% AT2_markers)

#write.xlsx(sd7_vs_o2d7_markers_sasp, file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/sd7_vs_o2d7_markers_sasp.xlsx', row.names = F)
#write.xlsx(sd7_vs_o2d7_markers_onecarbon , file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/sd7_vs_o2d7_markers_onecabon.xlsx', row.names = F)
#write.xlsx(sd7_vs_o2d7_markers_at2 , file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/sd7_vs_o2d7_markers_at2.xlsx', row.names = F)

#write.xlsx(sd7_vs_o2d7 , file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/sd7_vs_o2d7.xlsx', row.names = F)



sd7_vs_aird7_markers_sasp <- dplyr::filter(sd7_vs_aird7, gene %in% features_list )
sd7_vs_aird7_markers_onecarbon <- dplyr::filter(sd7_vs_aird7, gene %in% onecarbon_genes_fixed)
sd7_vs_aird7_markers_at2 <- dplyr::filter(sd7_vs_aird7, gene %in% AT2_markers)

write.xlsx(sd7_vs_aird7_markers_sasp, file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/sd7_vs_aird7_markers_sasp.xlsx', row.names = F)
write.xlsx(sd7_vs_aird7_markers_onecarbon , file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/sd7_vs_aird7_markers_onecabon.xlsx', row.names = F)
write.xlsx(sd7_vs_aird7_markers_at2 , file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/sd7_vs_aird7_markers_at2.xlsx', row.names = F)

write.xlsx(sd7_vs_aird7 , file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/sd7_vs_aird7.xlsx', row.names = F)



o2d_vs_aird7_markers_sasp <- dplyr::filter(o2d_vs_aird7, gene %in% features_list )
o2d_vs_aird7_markers_onecarbon <- dplyr::filter(o2d_vs_aird7, gene %in% onecarbon_genes_fixed)
o2d_vs_aird7_markers_at2 <- dplyr::filter(o2d_vs_aird7, gene %in% AT2_markers)

write.xlsx(o2d_vs_aird7_markers_sasp, file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/o2d_vs_aird7_markers_sasp.xlsx', row.names = F)
#write.xlsx(o2d_vs_aird7_markers_onecarbon , file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/o2d_vs_aird7_markers_onecabon.xlsx', row.names = F)
write.xlsx(o2d_vs_aird7_markers_at2 , file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/o2d_vs_aird7_markers_at2.xlsx', row.names = F)

write.xlsx(o2d_vs_aird7 , file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/o2d_vs_aird7.xlsx', row.names = F)
```


 
```{r "What's going on w the seurat cluster comparisons only in SD7 data?"}


SD7_cluster_markers

# Add the gene information to each cluster..
SD7_cluster_markers <- left_join(SD7_cluster_markers, all_genes_bm, by = c('gene' = 'mgi_symbol'))

#split up the markers by cluster..
cluster_levels <- unique(SD7_cluster_markers$cluster) 
SD7_cluster_markers.split <- lapply(seq_along(cluster_levels), function(i) SD7_cluster_markers %>% dplyr::filter(cluster == cluster_levels[i]))

#write.xlsx(SD7_cluster_markers, file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/SD7_cluster_markers.xlsx', row.names = F)

```

```{r "what about phagocytosis markers or AT2 markers only in SD7"}


SD7_cluster_markers_sasp <- dplyr::filter(SD7_cluster_markers, gene %in% features_list )

SD7_cluster_markers_at2 <- dplyr::filter(SD7_cluster_markers, gene %in% AT2_markers)

SD7_cluster_markers_phagocytosis <- SD7_cluster_markers %>% dplyr::filter(grepl('R-MMU-2029480', reactome_ids) | grepl('R-MMU-2029485', reactome_ids) | grepl('R-MMU-2029482', reactome_ids) | grepl('R-MMU-8941413', reactome_ids)) 


write.xlsx(SD7_cluster_markers_sasp, file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/SD7_cluster_markers_sasp.xlsx', row.names = F)

write.xlsx(SD7_cluster_markers_at2, file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/SD7_cluster_markers_at2.xlsx', row.names = F)

write.xlsx(SD7_cluster_markers_phagocytosis, file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/SD7_cluster_markers_phagocytosis.xlsx', row.names = F)
```


```{r "Look at the SASP genes in only sd7}

all_data.combined@meta.data$grouped_seurat <- paste0(all_data.combined@meta.data$atlas_predicted_clusters_grouped, "-", all_data.combined@meta.data$seurat_clusters)

Idents(all_data.combined) <- 'atlas_predicted_clusters_grouped'

sd7_cluster_markers_sasp_dotplot <- DotPlot(subset(all_data.combined, subset = orig.ident == 'SD7'), assay = 'SCT',features = features_list ) & 
  ggplot2::theme(axis.title = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.text=element_text(size=20), legend.text=element_text(size=20), legend.title=element_text(size=20)) & 
  coord_flip() & 
  scale_colour_gradient2(low = "blue", high = "red", midpoint = 0) &
  guides(color = guide_colorbar(title = 'Average Expression'))

ggsave('/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/sd7_cluster_markers_sasp_dotplot.png', plot = sd7_cluster_markers_sasp_dotplot, width = 12, height = 18)

Idents(all_data.combined) <- 'seurat_clusters'
library(RColorBrewer)

sd7_cluster_markers_sasp_dotplot_macrophages <-  DotPlot(subset(all_data.combined, subset = orig.ident == 'SD7' & atlas_predicted_clusters_grouped == 'Macrophages'), assay = 'SCT',features = features_list ) & 
  ggplot2::theme(axis.title = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.text=element_text(size=20), legend.text=element_text(size=20), legend.title=element_text(size=20)) &
  coord_flip() & 
  scale_colour_gradient2(low = "blue", high = "red", midpoint = 0) &
  guides(color = guide_colorbar(title = 'Average Expression'))
    

ggsave('/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/sd7_cluster_markers_sasp_dotplot_macrophages.png', plot = sd7_cluster_markers_sasp_dotplot_macrophages, width = 12, height = 18)



Idents(all_data.combined) <- 'orig.ident'


types_dotplot_ident <-  DotPlot(
  subset(all_data.combined, subset = (orig.ident == 'SD7' | orig.ident == 'O2D7' | orig.ident == 'AirD7')), assay = 'SCT',features = features_list) & ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) & 
  coord_flip()& 
  theme(axis.text=element_text(size=20))&
  scale_colour_gradient2(low = "blue", high = "red", midpoint = 0) &
  guides(color = guide_colorbar(title = 'Average Expression'))

ggsave('/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/types_dotplot_ident.png', plot = types_dotplot_ident , width = 12, height = 18)
```







DoHeatmap(subset(all_data.combined, subset = orig.ident == 'SD7'), assay = 'SCT',features = features_list )

macros <- sort(unique(all_data.combined@meta.data$grouped_seurat) %>% grep("Macrophages", ., value = TRUE))
at1 <- sort(unique(all_data.combined@meta.data$grouped_seurat) %>% grep("AT1", ., value = TRUE))
at2 <- sort(unique(all_data.combined@meta.data$grouped_seurat) %>% grep("AT2", ., value = TRUE))
endo <- sort(unique(all_data.combined@meta.data$grouped_seurat) %>% grep("Endo", ., value = TRUE))
mesen <- sort(unique(all_data.combined@meta.data$grouped_seurat) %>% grep("Mesenchymal", ., value = TRUE))
oth_immune <- sort(unique(all_data.combined@meta.data$grouped_seurat) %>% grep("Other immune cells", ., value = TRUE))
oth_epi <- sort(unique(all_data.combined@meta.data$grouped_seurat) %>% grep("Other epithelial cells", ., value = TRUE))

mca_seur_levels <- c(macros, at2, at1, mesen, endo, oth_epi, oth_immune)

all_data.combined@meta.data$grouped_seurat <- factor(all_data.combined@meta.data$grouped_seurat, levels = mca_seur_levels)
Idents(all_data.combined) <- 'seurat_clusters'

DotPlot(subset(all_data.combined, subset = orig.ident == 'SD7'), assay = 'SCT',features = features_list ) & ggplot2::theme(axis.title = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.text=element_text(size=20), legend.text=element_text(size=20), legend.title=element_text(size=20)) & coord_flip()


table(all_data.combined$seurat_clusters, all_data.combined$atlas_predicted_clusters_grouped)


Idents(all_data.combined) <- 'orig.ident'

orig_ident_sasp_dotplot <- DotPlot(
  subset(all_data.combined, subset = (orig.ident == 'SD7' | orig.ident == 'AirD7' | orig.ident == 'O2D7')), assay = 'SCT',features = features_list ) & ggplot2::theme(axis.title = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.text=element_text(size=20), legend.text=element_text(size=20), legend.title=element_text(size=20)) & coord_flip() & guides(colour = guide_legend(nrow = 2))

ggsave('/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/orig_ident_sasp_dotplot.png', plot = orig_ident_sasp_dotplot, width = 6, height = 12, dpi = 500)
```

```{r}
#which genes are unique markers for particular clusters?
table(SD7_cluster_markers_sasp$gene)
unique_sd7_sasp_genes <- c('Cdkn2c', 'Cdkn3', 'Cxcl1', 'Il1a', 'Il1b', 'Mmp11', 'Mmp23', 'Mmp8', 'Tnfsf14', 'Tnfsf9','Trp53i11')

SD7_cluster_markers_sasp_unique <- dplyr::filter(SD7_cluster_markers_sasp, gene %in% unique_sd7_sasp_genes )

write.xlsx(SD7_cluster_markers_sasp_unique, file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/SD7_cluster_markers_sasp_unique.xlsx', row.names = F)

```
```{r}
at2_marker_freqs <- data.frame(table(SD7_cluster_markers_at2$gene))
at2_marker_unique <- at2_marker_freqs %>% dplyr::filter(Freq == 1)
at2_marker_unique <- at2_marker_unique$Var1

SD7_cluster_markers_at2_unique <- dplyr::filter(SD7_cluster_markers_at2, gene %in% at2_marker_unique)
write.xlsx(SD7_cluster_markers_at2_unique, file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/SD7_cluster_markers_at2_unique.xlsx', row.names = F)

```





cluster_markers_split <- lapply(seq_along(cluster_levels), function(i) cluster_markers_ids %>% dplyr::filter(cluster == cluster_levels[i]))

cluster_levels <- levels(cluster_markers_ids$cluster)
cluster_markers_split <- lapply(seq_along(cluster_levels), function(i) cluster_markers_ids %>% dplyr::filter(cluster == cluster_levels[i]))
#names(cluster_markers_split) <- paste0('cluster_', cluster_levels, '_markers')


#doing some filtering to look at just the 'top markers', but probably want to only use these for making figures.
top5_markers <- SD7_seurat.markers_distinct  %>% group_by(cluster) %>% top_n(n = 5, wt = avg_logFC)

top5_markers.split <- lapply(seq_along(unique(top5_markers$cluster)), function(i) top5_markers %>% dplyr::filter(cluster == unique(top5_markers$cluster)[i]))
  
marker_dotplots <- lapply(seq_along(top5_markers.split), function(i) DotPlot(SD7_seurat, features = c((top5_markers.split[[i]])$gene)) & labs(title = paste0('seurat_', (top5_markers.split[[i]])$cluster)))


(marker_dotplots[[1]] + marker_dotplots[[2]] + marker_dotplots[[3]]) / (marker_dotplots[[4]] + marker_dotplots[[5]] + marker_dotplots[[6]]) 




SD7_cluster_markers_at2 %>% dplyr::filter(cluster == '17')
```



Idents(all_data.combined) <- 'atlas_predicted_clusters_short'

types_dotplot_sub_phagocytosis <-  DotPlot(subset(all_data.combined, subset = orig.ident == 'SD7'), assay = 'SCT',features = unique(SD7_seurat.markers_phagocytosis$gene) ) & ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) & coord_flip() & theme(axis.text=element_text(size=20), legend.text=element_text(size=20), axis.title = element_blank())

ggsave('/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/types_dotplot_sub_phagocytosis.png', plot = types_dotplot_sub_phagocytosis, width = 14, height = 18)

Idents(all_data.combined) <- 'seurat_clusters'

AT2_markers <- unique(c("Abhd5", "Acaa1a", "Adgrf5", "Aig1", "Arhgap44", "Atf3", "Atp13a3", "Bmp4", "Bnip3", "Ccz1", "Cdh16", "Cds1", "Cftr", "Clu", "Cndp2", "Col6a1", "Cracr2a", "Cyp51", "Dhcr24", "Dpp9", "Echs1", "Eif3b", "Epb41l5", "Erbb3", "Fam126b", "Fdft1", "Fgf1", "Foxp2", "Gata6", "Gsta4", "Hadh", "Hdc", "Higd1a", "Hmgcr", "Igsf3", "Il18r1", "Irs2", "Itih4", "Junos", "Kcne2", "Kcnj13", "Kcnj2", "Klf9", "LOC115487124", "Lratd2", "Maged1", "Malat1", "Mob3b", "Mrps9", "Myh7", "Nedd4l", "P4ha1", "Paqr8", "Pard3", "Pcnt", "Pcyt2", "Pls3", "Plxnb1", "Pof1b", "Ppp1r16a", "Rac3", "Rbms3", "Rn7sk", "Serpinb9", "Sgpp2", "Sh3d19", "Slc16a7", "Slc25a39", "Sowahc", "Spint1", "Spr", "Spred1", "Sqle", "Srebf2", "Steap2", "Syt7", "Tbc1d30", "Tcea3", "Tkt", "Tmc4", "Tmem205", "Tmem8", "Tmprss4", "Trim2", "Trp53inp2", "Trpm6", "Ttc38", "Txnrd3", "Usp33", "Vsig2","Rtkn2", "Spock2", "Col4a3", "2200002D01Rik", "Sema3e", "Lamc2", "Scnn1g", "Marco", "Itgb6", "Lama3", "Ccl6", "Igfbp7", "Fth1", "Fbln5", "Krt19", "Sec14l3", "Cd9", "Ano1", "Lpl", "Cryab", "F3", "Tmsb4x", "Vegfa", "Clic5", "S100a11", "Anxa2", "Top2a", "Myl6", "Col4a1", "Actb", "Hs2st1", "Agrn", "Ear2", "Plet1", "Fam174b", "Atp6v0d2", "Fzd2", "Nnat", "Pclaf", "Hspg2", "Rpl36al", "Mest", "Rps27a", "Ctsd", "Cstb", "Akr1a1", "Col4a2", "Calm1", "Fads3", "Plin2", "Tspan15", "Ndufa4", "Ndufb1-ps", "Rpl19", "Aqp1", "Rpl26", "Rpl11", "Ywhae", "Cox4i1", "Sem1", "Anxa5", "Fam189a2", "Uqcrb", "Eif1", "Rps20", "Rpl35", "Calm2", "Ftl1", "Ndst1", "Atp5l", "Cox8a", "Col4a5", "H3f3b", "Serpinh1", "Btf3", "Serf2", "Rpl34", "Tmsb10", "Itgb1", "Cox5b", "Tagln2", "Rpl13", "Rpl14", "Rpl22l1", "Rpl36", "Hsp90ab1", "Trp53bp2", "Myl12b", "Atp5j", "Rab11fip1", "Ctsb", "Lamp1", "Ptov1", "Rps4x", "Rpl22", "Atp5pb", "Ctsz", "Rps13", "Ehd2", "Rpl35a", "Hspa8", "Rps24", "S100a1", "Rps3a1", "Lrpap1", "Drap1", "Rhoa", "Pdia3", "Ndufa13", "Atp5md", "Cd81", "Cox5a", "F11r", "Rpl7", "Cttnbp2nl", "Mrpl14", "Eif2s2", "Ybx1", "1810037I17Rik", "Arf1", "Rpl37a", "Rps23", "Tppp3", "Tmem176b", "Atp5mpl", "Rpl18", "Lamc1", "Rpl36a", "Atp5b", "Npm1", "Ahnak", "Tomm6", "Rpl27", "Rpl9", "Anxa1", "Rpl23", "Rpl8", "Mmp14", "Car2", "Car2", "Rps3", "Canx", "Rps15a", "Xist", "Paip1", "Atp5g1", "Rpl27a", "Tomm7", "Uba52", "Rps27l", "Myof", "H3f3a", "Hspa5", "Slc25a4", "Pmp22", "Tubb4b", "Tmed10", "Hmgb2", "Gng5", "Smdt1", "Atp5c1", "Laptm4a", "Spcs2", "Edf1", "Ctdspl", "Eef1a1", "Eif4a1", "Mpc2", "Crip1", "Slc25a3", "Rpl37", "Rps8", "Ubl5", "Rpl38", "Gnb1", "Rpl28", "Ndufb9", "Gnb2", "LOC102636299", "Cyba", "Qk", "Eif3f", "Rpl30", "Dad1", "Eef2", "Esd", "Pcna", "Pfn1", "Ndufa5", "Hsp90b1", "Actn4", "Uqcrq", "Tuba1a", "Ndufa3", "Cnn2", "Tmem256", "Galnt18", "Ndufb7", "Uqcr10", "Snrpe", "Cux1", "Ndufb8", "Sptbn1", "Atp5g3", "Arf5", "Ndufa8", "Hnrnpa3", "Rpl37rt", "Selenow", "Ndufa1", "Psap", "Prdx5", "Rps9", "Ost4", "Atp5o", "Atp6v0d1", "Selenoh", "Clta", "Reep5", "Mrfap1", "Ndufb5", "Pon2", "Hypk", "Pomp", "Sox4", "Mrps21", "Ssr4", "Dctn3", "Tpm3", "Rpl17", "Fdx1", "Rps15a-ps6", "Aldoa", "Hsbp1", "Ndufb2", "Cox7b", "Fermt2", "Serbp1", "Acaa2", "Psmb2", "Nedd8", "Morf4l1", "Uqcrh", "Ndufc1", "Cdk4", "Hnrnpk", "Abcg1", "Ppp2cb", "Cycs", "Vdac2", "Ndufa2", "Tjp1", "Fis1", "Ran", "Cavin1", "Cdc42", "Aprt", "Eif3h", "Romo1", "Atp6v1f", "Ywhah", "Snrpd2", "Cfl1", "Tma7", "Slc39a1", "Nme1", "Cbx3", "Rbx1", "Micos10", "Ndufb11", "Rtraf", "Krtcap2", "Ubc", "Ifi27", "Ostc", "Alyref", "Mlf2", "Arhgap5", "Rab5if", "Dynlrb1", "Svil", "Lamtor2", "Tbca", "Snrpg", "Ctnna1", "Psma7", "Dpysl2", "Psmb1", "Fkbp1a", "1810058I24Rik", "Itm2c", "Hdgf", "Plp2", "Hadha", "Ndufs6", "Hmgn2", "Capns1", "D8Ertd738e", "Tpm4", "Gm9385", "Swi5", "Pdcd5", "Pebp1", "Anp32a", "Mpc1", "Tomm20", "Atp6v0e", "Erh", "Itm2b", "Mif", "Arpp19", "Psma2", "Cox17", "Psmb3", "Tmed2", "Skp1a", "Ywhaq", "Vcp", "Spcs1", "Kdelr2", "Psmd8", "Ube2d3", "Hnrnpa0", "Ndufa11", "Pkm", "Dync1i2", "Emd", "Psmb5", "Bzw1", "Tmem258", "Hnrnpm", "Rhob", "Hnrnpd", "Selenok", "Alas2", "Alas2", "Igfbp5", "Ube2c", "Gpc3", "Bpgm", "Bpgm", "Bpgm", "Bnip3l"))

SD7_seurat.markers_at2 <- dplyr::filter( SD7_seurat.markers, gene %in% AT2_markers)

write.xlsx(SD7_seurat.markers_at2, file = '/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/SD7_cluster_markers_at2.xlsx', row.names = F)

SD7_seurat.markers_at2$gene

Idents(all_data.combined) <- 'atlas_predicted_clusters_short'

types_dotplot_sub_at2 <-  DotPlot(subset(all_data.combined, subset = orig.ident == 'SD7'), assay = 'SCT',features = unique(SD7_seurat.markers_at2$gene) ) & ggplot2::theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) & coord_flip() & theme(axis.text=element_text(size=20), legend.text=element_text(size=20), axis.title = element_blank())

ggsave('/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/types_dotplot_sub_at2.png', plot = types_dotplot_sub_at2, width = 14, height = 18)

```

```{r}
#install.packages("writexl")
library(xlsx)

sheet_names <- c(paste0('cluster', '_', cluster_levels))

names(SD7_seurat.markers.split) <- paste0('seurat_', unique(top5_markers$cluster))

write.xlsx(SD7_seurat.markers.split, "/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/markers_per_cluster_SD7.xlsx")

#doing some filtering to look at just the 'top markers', but probably want to only use these for making figures.
top5_markers <- SD7_seurat.markers_distinct  %>% group_by(cluster) %>% top_n(n = 5, wt = avg_logFC)

top5_markers.split <- lapply(seq_along(unique(top5_markers$cluster)), function(i) top5_markers %>% dplyr::filter(cluster == unique(top5_markers$cluster)[i]))
  
marker_dotplots <- lapply(seq_along(top5_markers.split), function(i) DotPlot(SD7_seurat, features = c((top5_markers.split[[i]])$gene)) & labs(title = paste0('seurat_', (top5_markers.split[[i]])$cluster)))


(marker_dotplots[[1]] + marker_dotplots[[2]] + marker_dotplots[[3]]) / (marker_dotplots[[4]] + marker_dotplots[[5]] + marker_dotplots[[6]]) 

#ggsave(filename = 'marker_dotplots.png', width = 30, height = 30, device = 'png', plot = (marker_dotplots[[1]] + marker_dotplots[[2]] + marker_dotplots[[3]] + marker_dotplots[[4]] + marker_dotplots[[5]] + marker_dotplots[[6]] + marker_dotplots[[7]] + marker_dotplots[[8]] + marker_dotplots[[9]] + marker_dotplots[[10]] + 
#marker_dotplots[[11]] + marker_dotplots[[12]] + marker_dotplots[[13]] + marker_dotplots[[14]] + marker_dotplots[[15]] + marker_dotplots[[16]] + marker_dotplots[[17]] + marker_dotplots[[18]] + marker_dotplots[[19]] + marker_dotplots[[20]] + marker_dotplots[[21]] + marker_dotplots[[22]]))


DefaultAssay(SD7_seurat) <- 'SCT'

marker_dotplots <-  DotPlot(SD7_seurat, features = top5_markers$gene, group.by = 'seurat_clusters') & 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) & 
  geom_vline(xintercept = 5.55) &
  geom_vline(xintercept = 10.55) &
  geom_vline(xintercept = 15.55) &
  geom_vline(xintercept = 20.55) &
  geom_vline(xintercept = 25.55) &
  geom_vline(xintercept = 30.55) &
  geom_vline(xintercept = 35.55) &
  geom_vline(xintercept = 40.55) &
  geom_vline(xintercept = 45.55) &
  geom_vline(xintercept = 50.55) &
  geom_vline(xintercept = 55.55) &
  geom_vline(xintercept = 60.55) &
  geom_vline(xintercept = 65.55) &
  geom_vline(xintercept = 70.55) &
  geom_vline(xintercept = 75.55) &
  geom_vline(xintercept = 80.55) &
  geom_vline(xintercept = 85.55) &
  geom_vline(xintercept = 90.55) &
  geom_vline(xintercept = 95.55) &
  geom_vline(xintercept = 100.55) &
  geom_vline(xintercept = 105.55) &
  geom_vline(xintercept = 110.55) &
  geom_hline(yintercept = 1.5) &
  geom_hline(yintercept = 2.5) &
  geom_hline(yintercept = 3.5) &
  geom_hline(yintercept = 4.5) &
  geom_hline(yintercept = 5.5) &
  geom_hline(yintercept = 6.5) &
  geom_hline(yintercept = 7.5) &
  geom_hline(yintercept = 8.5) &
  geom_hline(yintercept = 9.5) &
  geom_hline(yintercept = 10.5) &
  geom_hline(yintercept = 11.5) &
  geom_hline(yintercept = 12.5) &
  geom_hline(yintercept = 13.5) &
  geom_hline(yintercept = 14.5) &
  geom_hline(yintercept = 15.5) &
  geom_hline(yintercept = 16.5) &
  geom_hline(yintercept = 17.5) &
  geom_hline(yintercept = 18.5) &
  geom_hline(yintercept = 19.5) &
  geom_hline(yintercept = 20.5) &
  geom_hline(yintercept = 21.5) &
  geom_hline(yintercept = 22.5) &
  geom_hline(yintercept = 23.5) &
  geom_hline(yintercept = 24.5) &
  geom_hline(yintercept = 25.5) &
  geom_hline(yintercept = 26.5) &
  geom_hline(yintercept = 27.5) &
  geom_hline(yintercept = 28.5) &
  geom_hline(yintercept = 29.5) &
  geom_hline(yintercept = 30.5) &
  geom_hline(yintercept = 31.5) &
  geom_hline(yintercept = 32.5) &
  geom_hline(yintercept = 33.5)


#Results are saved in a new assay (named SCT by default) with counts being (corrected) counts, data being log1p(counts), scale.data being pearson residuals; sctransform::vst intermediate results are saved in misc slot of new assay. 

FeaturePlot(SD7_seurat, 
            label = TRUE, 
            label.size = 6, 
            repel = F, 
            sort = TRUE, 
            features = (top5_markers %>% dplyr::filter(cluster == 0))$gene, 
            ncol = 5, 
            reduction = 'tsne',
            slot = 'counts') &
  scale_colour_gradientn(limits= c(0,8), colors = c('bisque', 'darkred')) &
  labs(subtitle = 'SD7')


DotPlot(SD7_seurat, 
        assay = 'SCT',
        features = (top5_markers %>% dplyr::filter(cluster == 8))$gene) & 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) & 
  coord_flip()
```

```{r "Table comparing the numbers of cells in each combination of seurat cluster and cell type annotation"}
cell_types_per_cluster_df  <- data.frame(atlas = SD7_seurat$atlas_predicted_clusters_short, seurat = SD7_seurat$seurat_clusters)
cell_types_per_cluster <- as.data.frame(table(cell_types_per_cluster_df$atlas, cell_types_per_cluster_df$seurat))
cell_types_per_cluster$atlas <- cell_types_per_cluster$Var1
cell_types_per_cluster$seurat <- cell_types_per_cluster$Var2
cell_types_per_cluster$Var1 <- NULL
cell_types_per_cluster$Var2 <- NULL
cell_types_per_cluster <- cell_types_per_cluster %>% arrange(desc(Freq))
cell_types_per_cluster <- cell_types_per_cluster %>% filter(Freq > 0)
cell_types_per_cluster <- cell_types_per_cluster %>% arrange(desc(atlas))
#write.csv(cell_types_per_cluster, 'cell_types_per_cluster.csv', row.names= F)

cell_types_per_cluster$number_of_cells <- cell_types_per_cluster$Freq
cell_types_per_cluster$Freq <- NULL
cell_types_per_cluster_plot <- ggplot(cell_types_per_cluster, aes(x = atlas, y = seurat, fill = number_of_cells)) +
  coord_equal()+
  geom_tile(colour = 'black') + 
  xlab('Cell Annotation') +
  ylab('Seurat Cluster') +
  theme(panel.grid = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank()) +
  scale_fill_gradientn(
    colors = c("white", "darkred"),
    limits = range(cell_types_per_cluster$number_of_cells)) +
  theme(panel.grid = element_blank(),
        axis.text.x=element_text(angle=90, hjust = 1, vjust = .4),
        axis.ticks = element_blank(),
        text = element_text(size=20)) 
```

```{r "Table of how many cells in each cell type category"}
cell_types_freq_df <- data.frame(ident = SD7_seurat$orig.ident, atlas = SD7_seurat$atlas_predicted_clusters_short)
cell_types_freq <- as.data.frame(table(cell_types_freq_df$ident, cell_types_freq_df$atlas))
cell_types_freq$atlas <- cell_types_freq$Var2
cell_types_freq$Var2 <- NULL
cell_types_freq$ident <- cell_types_freq$Var1
cell_types_freq$Var1 <- NULL

cell_types_freq <- cell_types_freq %>% arrange(desc(Freq))
cell_types_freq <- cell_types_freq %>% filter(Freq > 0)
#write.csv(cell_types_freq, 'cell_types_freq.csv', row.names= F)


cell_types_freq$number_of_cells <- cell_types_freq$Freq
cell_types_freq$Freq <- NULL

cell_types_freq_plot <- ggplot(cell_types_freq, aes(x = ident, y = atlas, fill = number_of_cells)) +
  coord_equal()+
  geom_tile(colour = 'black') + 
  xlab('Ident') +
  ylab('Cell state') +
  theme(panel.grid = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank()) +
  scale_fill_gradientn(
    colors = c("white", "red"),
    limits = range(cell_types_freq$number_of_cells)) +
  geom_text(aes(label = number_of_cells)) +
  theme(panel.grid = element_blank(),
        axis.text.x=element_text(angle=90, hjust = 1, vjust = .4),
        axis.ticks = element_blank(),
        text = element_text(size=20))

```

```{r "Table of how many cells are in each seurat cluster"}
clusters_freq_df <- data.frame(ident = SD7_seurat$orig.ident, seurat = SD7_seurat$seurat_clusters)
clusters_freq <- as.data.frame(table(clusters_freq_df$ident, clusters_freq_df$seurat))
clusters_freq$seurat <- clusters_freq$Var2
clusters_freq$Var1 <- NULL
clusters_freq$Var2 <- NULL
clusters_freq <- clusters_freq %>% arrange(desc(Freq))
clusters_freq$Number_of_cells <- clusters_freq$Freq
#write.csv(clusters_freq, 'clusters_freq.csv', row.names= F)

#Are there amy mesenchymal cells in the SD7?

cell_types_freq_df <- data.frame(ident = all_data.combined$orig.ident, atlas = all_data.combined$atlas_predicted_clusters_short)
cell_types_freq <- as.data.frame(table(cell_types_freq_df$ident, cell_types_freq_df$atlas))
cell_types_freq$atlas <- cell_types_freq$Var2
cell_types_freq$Var2 <- NULL
cell_types_freq$ident <- cell_types_freq$Var1
cell_types_freq$Var1 <- NULL

cell_types_freq <- cell_types_freq %>% arrange(desc(Freq))
cell_types_freq <- cell_types_freq %>% filter(Freq > 0)


cell_types_freq$number_of_cells <- cell_types_freq$Freq
cell_types_freq$Freq <- NULL

cell_types_freq_plot <- ggplot(cell_types_freq, aes(x = ident, y = atlas, fill = number_of_cells)) +
  coord_equal()+
  geom_tile(colour = 'black') + 
  xlab('Ident') +
  ylab('Cell state') +
  theme(panel.grid = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank()) +
  scale_fill_gradientn(
    colors = c("white", "red"),
    limits = range(cell_types_freq$number_of_cells)) +
  geom_text(aes(label = number_of_cells)) +
  theme(panel.grid = element_blank(),
        axis.text.x=element_text(angle=90, hjust = 1, vjust = .4),
        axis.ticks = element_blank(),
        text = element_text(size=20))
        cell_types_freq_plot


cell_types_freq_plot        



cell_types_freq_plot_sd7 <- ggplot(cell_types_freq %>% dplyr::filter(ident == 'SD7'), 
aes(x = ident, y = atlas, fill = number_of_cells)) +
  coord_equal()+
  geom_tile(colour = 'black') + 
  xlab('Ident') +
  ylab('Cell state') +
  theme(panel.grid = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank()) +
  scale_fill_gradientn(
    colors = c("white", "red"),
    limits = range(
    (
    cell_types_freq %>% dplyr::filter(ident == 'SD7')
    )$number_of_cells)) +
  geom_text(aes(label = number_of_cells)) +
  theme(panel.grid = element_blank(),
        axis.text.x=element_text(angle=90, hjust = 1, vjust = .4),
        axis.ticks = element_blank(),
        text = element_text(size=20))

cell_types_freq_plot_sd7


#ggsave('/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/cell_types_freq_plot.png', plot = cell_types_freq_plot, width = 14, height = 14)

#ggsave('/gpfs/data/cbc/Dennery_scRNAseq/scRNAseq/Dennery_scRNA/scRNAseq/cell_types_freq_plot_sd7_only.png', plot = cell_types_freq_plot_sd7, width = 14, height = 14)

#ggsave('cell_types_per_cluster_plot.png', device = 'png')
#ggsave(plot = conc_plot2, file = 'figure2b1.png', width = 10, height = 15, units = 'in')

```



```{r}
#all_genes <- data.frame(rownames(all_data.combined)) %>% rename(rownames.all_data.combined. = "mgi_symbol")
mouse_biomart <- useMart("ensembl", dataset = "mmusculus_gene_ensembl")

mouse_biomart_data <- getBM(mart = mouse_biomart, attributes=c("mgi_description","mgi_symbol", "kegg_enzyme", "name_1006"))


all_genes_bm <- right_join(mouse_biomart_data, all_genes)
```






# AT1, AT2, and B cells are in similar clusters (does that make sense? are they similar cell types?)
# AT1 cells are in cluster 25, top markers: Ndst1, Mal2, Tead1, Agrn, Dag1, between 35 and 52 cells in each experiment
# AT2 cells are in a few clusters -- 3, 11, 13, 22, and 27. 22 and 27 have other cell types too (B cells, AT1 cells, alveolar macrophages), but are mainly AT2.
# Cluster 3 markers: Cat, Irx3, Pla2g1b, Ctsh, Elf5
# Cluster 11 markers: Kcnc3, LOC115487417, Ppp1r9a, Gm33122, St6galnac2
# Cluster 13 markers: Fasn, Tgoln1, Il33, Lrg2, Lrp2
# Cluster 22 markers: 



```{r How many cells are in each cell type?}
DefaultAssay(all_data.combined) <- 'SCT'

all_data.combined_meta.data <- tibble(all_data.combined@meta.data) 

counts_per_cluster_and_ident <- all_data.combined_meta.data %>% group_by(seurat_clusters, orig.ident) %>% tally() 
counts_per_cluster_and_ident_spread <- counts_per_cluster_and_ident %>% tidyr::spread(key=orig.ident, value=n)
nrow(all_data.combined_meta.data)
colSums(Filter(is.numeric, counts_per_cluster_and_ident_spread), na.rm = T)


counts_per_annot_and_ident <- all_data.combined_meta.data %>% group_by(predicted.id, orig.ident) %>% tally() 
counts_per_annot_and_ident_spread <- counts_per_annot_and_ident %>% tidyr::spread(key=orig.ident, value=n)
nrow(all_data.combined_meta.data)
colSums(Filter(is.numeric, counts_per_annot_and_ident_spread), na.rm = T)

table(all_data.combined@meta.data$seurat_clusters, all_data.combined@meta.data$orig.ident)
table(all_data.combined@active.ident, all_data.combined@meta.data$orig.ident)
table(counts_per_annot_and_ident_spread)

#https://github.com/satijalab/seurat/issues/738

count_plot_seurat <- ggplot(counts_per_cluster_and_ident, aes(x = orig.ident, y = seurat_clusters, fill = n)) + 
  geom_tile() +
  theme(panel.grid = element_blank(),
        axis.title.y = element_blank(),
        axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank()) +
  geom_text(aes(label = n)) +
  xlab("") +
  scale_fill_gradientn(
    colors = c("white", "steelblue3"),
    limits = range(counts_per_cluster_and_ident$n)) +
  theme_minimal()

count_plot_mca <- ggplot(counts_per_annot_and_ident, aes(x = orig.ident, y = predicted.id, fill = n)) + 
  geom_tile() +
  theme(panel.grid = element_blank(),
        axis.title.y = element_blank(),
        axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank()) +
  geom_text(aes(label = n)) +
  xlab("") +
  scale_fill_gradientn(
    colors = c("white", "steelblue3"),
    limits = range(counts_per_annot_and_ident$n)) +
  theme_minimal()

#write.table((table(all_data.combined@meta.data$seurat_clusters, all_data.combined@meta.data$orig.ident)), "/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/cells_per_seurat_cluster.txt", sep = '\t')

#write.table((table(all_data.combined@meta.data$predicted.id, all_data.combined@meta.data$orig.ident)), "/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/cells_per_mca_cluster_annot.txt", sep = '\t')
```



```{r}
#saveRDS(all_data.combined, file = "/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/all_data.combined.RDS")
#all_data.combined <- readRDS(file = "/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/all_data.combined.RDS")
```

```{r Differential Expression}
# It seems like folks are interested in a few clusters:
# Cluster 25 - AT1 cells
# Clusters 3, 11, 13, 22, and 27 are the AT2 cell types (22 and 27 have other cell types too).
# Ciliated cells are in Cluster 28
# Club cells are in Cluster 39

levels(all_data.combined$orig.ident)

#make a new column in all_data.combined@meta.data that combines the seurat cluster and orig.ident
all_data.combined@meta.data$ident_clust <- paste0(all_data.combined@meta.data$orig.ident, "-", all_data.combined@meta.data$seurat_clusters)

all_data.combined@meta.data$treat <- tidyr::separate(data = all_data.combined@meta.data, 
                col = orig.ident,
                into = c('treat','time'),
                sep = 'D',
                remove = F)$treat


all_data.combined@meta.data$time <- tidyr::separate(data = all_data.combined@meta.data, 
                col = orig.ident,
                into = c('treat','time'),
                sep = 'D',
                remove = F)$time
```

```{r Look at marker genes across clusters e.g., which genes are most important for distinguishing the cell clusters -- we were using the integrated data previously, but I think using the 'SCT' normalized data is a better idea.}
Idents(all_data.combined) <- all_data.combined$seurat_clusters
DefaultAssay(all_data.combined) <- 'SCT'

# find markers for every cluster compared to all remaining cells, report only the positive ones
all_data.combined.markers.sct <- FindAllMarkers(all_data.combined, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25, assay = 'SCT')

# Add the gene information to each cluster..
all_data.combined.markers.sct <- left_join(all_data.combined.markers.sct, all_genes_bm, by = c('gene' = 'mgi_symbol'))

#split up the markers by cluster..
cluster_levels <- levels(all_data.combined.markers.sct$cluster)
all_data.combined.markers.sct.split <- lapply(seq_along(cluster_levels), function(i) all_data.combined.markers.sct %>% dplyr::filter(cluster == cluster_levels[i]))

per_cluster_markers <- all_data.combined.markers.sct.split

#saveRDS(per_cluster_markers, '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/per_cluster_markers.Rds')

#per_cluster_markers <- readRDS('/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/per_cluster_markers.Rds')
```


```{r}
Idents(all_data.combined) <- all_data.combined$orig.ident
DefaultAssay(all_data.combined) <- 'SCT'

# This chunk will take the Seurat object as an input and filter is such that each seurat cluster is in its own seurat object and then findallmarkers, then make dotplots of the top markers across the experimental conditions

#set the levels
clusters <- levels(all_data.combined@meta.data$seurat_clusters)

construct <- unlist(lapply(seq_along(clusters), function(i) paste0("all_data.combined_", clusters[i], "<- subset(all_data.combined, subset = seurat_clusters =='", clusters[i], "')")))

seurat.list <- lapply(seq_along(construct), function(i) eval(parse(text = construct[i])))

names(seurat.list) <- c(levels(all_data.combined@meta.data$seurat_clusters))

seurat.list.markers <- lapply(seurat.list, FindAllMarkers, min.pct = 0.25, logfc.threshold = 0.25, assay = 'SCT')

#Add the gene explanations to each file..
seurat.list.markers.ids <- lapply(seq_along(seurat.list.markers), function(i) left_join(seurat.list.markers[[i]], all_genes_bm, by = c('gene' = 'mgi_symbol')))

per_condition_markers <- seurat.list.markers.ids
#saveRDS(per_condition_markers, '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/per_condition_markers.Rds')

names(per_condition_markers) <- names(seurat.list)

#write the findallmarkers outputs for each cluster to a file
#lapply(seq_along(seurat.list.markers.ids), function(i) write.csv(seurat.list.markers.ids[[i]], file = paste0("/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/figures_and_tables/tables/condition_markers/cluster_", i-1, "_per_condition_markers.csv"), row.names = F))

```
```{r}
DefaultAssay(all_data.combined) <- 'SCT'
FeaturePlot(all_data.combined, features = 'Malat1', split.by = 'orig.ident', reduction = 'tsne', slot = 'data', cols = )

```



```{r Re-import the per-condition markers and try to collate them into one table that focuses on the O2D60 data}
#first, add a column to each object in the list, the column should just be its own index or name
per_condition_markers_bind <- rbindlist(per_condition_markers, idcol = "seurat_cluster")

per_condition_markers_bind_sort <- per_condition_markers_bind %>% dplyr::arrange(factor(cluster, levels = c('O2D60','AirD60', 'O2D7', 'AirD7')))

per_condition_markers_bind_sort <- dplyr::left_join(per_condition_markers_bind_sort, all_genes_bm, by = c('gene' = 'mgi_symbol'))

#write.csv(per_condition_markers_bind_sort, file = '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/per_condition_markers_bind_sort.csv', row.names = F)

#write.xlsx(per_condition_markers_bind_sort, file = '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/per_condition_markers_bind_sort.xlsx', row.names = F)
```



```{r Looking at overall condition markers (rather than per-cluster), looking at featureplots..}

Idents(all_data.combined) <- all_data.combined$orig.ident
DefaultAssay(all_data.combined) <- 'SCT'

all_condition_markers <- FindAllMarkers(all_data.combined, min.pct = 0.25, logfc.threshold = 0.25, assay = 'SCT')

all_condition_markers_sort <- all_condition_markers %>% dplyr::arrange(factor(cluster, levels = c('O2D60','AirD60', 'O2D7', 'AirD7')))
all_condition_markers_sort %>% dplyr::filter(cluster == 'O2D60')

Idents(all_data.combined) <- all_data.combined$seurat_clusters
  
features <- c("Cftr", "Spock2", "Malat1", "Tkt", "Alas1")
DotPlot(all_data.combined, features = features, group.by = 'seurat_clusters', assay = 'SCT') + RotatedAxis()
    
rownames(all_data.combined@assays$SCT@counts)

all_data.combined$seurat_clusters
```


```{r}

get_top_markers <- function(seurat_markers){
  output <- seurat_markers %>% group_by(cluster)  %>% top_n(n = 10, wt = avg_logFC)
  return(output)
}

seurat.list.top.markers <- lapply(seurat.list.markers, get_top_markers)

#seurat_list_dotplots <- lapply(seq_along(seurat.list), function(i) DotPlot(seurat.list[[i]], features = c(unique(seurat.list.top.markers[[i]]$gene))) + RotatedAxis() + ggtitle(paste0("Seurat Cluster ", names(seurat.list[i]))))

#lapply(seq_along(seurat_list_dotplots), function(i) ggsave(plot =) seurat_list_dotplots[[i]], file = paste0("/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/figures/cluster_", i-1, "_dotplot.pdf"), width = 11,  height = 7))
```

```{r}
#3, 11, 13, 22, 27

cluster_3_markers <- per_cluster_markers[[4]]
cluster_11_markers <- per_cluster_markers[[12]]
cluster_13_markers <- per_cluster_markers[[14]]
cluster_22_markers <- per_cluster_markers[[23]]
cluster_27_markers <- per_cluster_markers[[28]]
```

```{r}
#make some lists..
cluster_3_compare <- cluster_11_markers %>%
  full_join(cluster_13_markers, by = 'gene') %>%
  full_join(cluster_22_markers, by = 'gene') %>%
  full_join(cluster_27_markers, by = 'gene')

cluster_11_compare <- cluster_3_markers %>%
  full_join(cluster_13_markers, by = 'gene') %>%
  full_join(cluster_22_markers, by = 'gene') %>%
  full_join(cluster_27_markers, by = 'gene')

cluster_13_compare <- cluster_11_markers %>%
  full_join(cluster_3_markers, by = 'gene') %>%
  full_join(cluster_22_markers, by = 'gene') %>%
  full_join(cluster_27_markers, by = 'gene')

cluster_22_compare <- cluster_11_markers %>%
  full_join(cluster_3_markers, by = 'gene') %>%
  full_join(cluster_13_markers, by = 'gene') %>%
  full_join(cluster_27_markers, by = 'gene')

cluster_27_compare <- cluster_11_markers %>%
  full_join(cluster_3_markers, by = 'gene') %>%
  full_join(cluster_13_markers, by = 'gene') %>%
  full_join(cluster_22_markers, by = 'gene')

#markers in cluster 3 but not in 11, 13, 22, or 27
cluster_3_specific <- anti_join(cluster_3_markers, cluster_3_compare, by="gene") 
cluster_11_specific <- anti_join(cluster_11_markers, cluster_11_compare, by="gene") 
cluster_13_specific <- anti_join(cluster_13_markers, cluster_13_compare, by="gene") 
cluster_22_specific <- anti_join(cluster_22_markers, cluster_22_compare, by="gene") 
cluster_27_specific <- anti_join(cluster_27_markers, cluster_27_compare, by="gene") 

```

```{r}
#Compare to supplemental data
supp_table<-read.csv('/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/41586_2018_BFnature25786_MOESM2_ESM.csv')
supp_table <- left_join(supp_table, all_genes_bm, by = c('SYMBOL' = 'mgi_symbol'))
```

```{r}
cluster_3_specific_and_supp <- inner_join(supp_table, cluster_3_specific, by = c('SYMBOL' = 'gene')) 
cluster_13_specific_and_supp <- inner_join(supp_table, cluster_13_specific, by = c('SYMBOL' = 'gene'))
cluster_11_specific_and_supp <- inner_join(supp_table, cluster_11_specific, by = c('SYMBOL' = 'gene'))
cluster_22_specific_and_supp <- inner_join(supp_table, cluster_22_specific, by = c('SYMBOL' = 'gene'))
cluster_27_specific_and_supp <- inner_join(supp_table, cluster_27_specific, by = c('SYMBOL' = 'gene'))

#write.csv(cluster_3_specific_and_supp, '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/Comparing_clust_3_11_13_22_27/cluster_3_specific_and_supp.csv', row.names = F)
#write.csv(cluster_11_specific_and_supp, '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/Comparing_clust_3_11_13_22_27/cluster_11_specific_and_supp.csv', row.names = F)
#write.csv(cluster_13_specific_and_supp, '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/Comparing_clust_3_11_13_22_27/cluster_13_specific_and_supp.csv', row.names = F)
#write.csv(cluster_22_specific_and_supp, '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/Comparing_clust_3_11_13_22_27/cluster_22_specific_and_supp.csv', row.names = F)
#write.csv(cluster_27_specific_and_supp, '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/Comparing_clust_3_11_13_22_27/cluster_27_specific_and_supp.csv', row.names = F)

#write.csv(cluster_3_specific, '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/Comparing_clust_3_11_13_22_27/cluster_3_specific.csv', row.names = F)
#write.csv(cluster_11_specific, '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/Comparing_clust_3_11_13_22_27/cluster_11_specific.csv', row.names = F)
#write.csv(cluster_13_specific, '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/Comparing_clust_3_11_13_22_27/cluster_13_specific.csv', row.names = F)
#write.csv(cluster_22_specific, '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/Comparing_clust_3_11_13_22_27/cluster_22_specific.csv', row.names = F)
#write.csv(cluster_27_specific, '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/Comparing_clust_3_11_13_22_27/cluster_27_specific.csv', row.names = F)

```

```{r}
#Let's also look at all the markers in clusters 3, 11, 13, 22, and 27 -- is there any KEGG enrichment there? 
library(clusterProfiler)
install.packages('ggnewscale')

#get the Ensembl mart:
mouse_biomart <- useMart("ensembl", 
                         dataset = "mmusculus_gene_ensembl")

entrez_universe <- getBM(mart = mouse_biomart, 
                                attributes = c("mgi_description", "mgi_symbol", "entrezgene_id"), 
                                values = rownames(all_data.combined), 
                                filters = "mgi_symbol")

entrez_universe$entrezgene_id < as.character(entrez_universe$entrezgene_id)
```

```{r}

mouse_biomart_entrez_3 <- getBM(mart = mouse_biomart, 
                                attributes = c("mgi_description", "mgi_symbol", "entrezgene_id"), 
                                values = cluster_3_specific$gene, 
                                filters = "mgi_symbol")
gseKEGG_3 <- enrichKEGG(gene = mouse_biomart_entrez_3$entrezgene_id, 
                        organism = 'mmu',
                        universe = as.character(entrez_universe$entrezgene_id))


mouse_biomart_entrez_11 <- getBM(mart = mouse_biomart, 
                                attributes = c("mgi_description", "mgi_symbol", "entrezgene_id"), 
                                values = cluster_11_specific$gene, 
                                filters = "mgi_symbol")
gseKEGG_11 <- enrichKEGG(gene = mouse_biomart_entrez_11$entrezgene_id, 
                        organism = 'mmu',
                        universe = as.character(entrez_universe$entrezgene_id))


mouse_biomart_entrez_13 <- getBM(mart = mouse_biomart, 
                                attributes = c("mgi_description", "mgi_symbol", "entrezgene_id"), 
                                values = cluster_13_specific$gene, 
                                filters = "mgi_symbol")
gseKEGG_13 <- enrichKEGG(gene = mouse_biomart_entrez_13$entrezgene_id, 
                        organism = 'mmu',
                        universe = as.character(entrez_universe$entrezgene_id))


mouse_biomart_entrez_22 <- getBM(mart = mouse_biomart, 
                                attributes = c("mgi_description", "mgi_symbol", "entrezgene_id"), 
                                values = cluster_22_specific$gene, 
                                filters = "mgi_symbol")
gseKEGG_22 <- enrichKEGG(gene = mouse_biomart_entrez_22$entrezgene_id, 
                        organism = 'mmu',
                        universe = as.character(entrez_universe$entrezgene_id))


mouse_biomart_entrez_27 <- getBM(mart = mouse_biomart, 
                                attributes = c("mgi_description", "mgi_symbol", "entrezgene_id"), 
                                values = cluster_27_specific$gene, 
                                filters = "mgi_symbol")
gseKEGG_27 <- enrichKEGG(gene = mouse_biomart_entrez_27$entrezgene_id, 
                        organism = 'mmu',
                        universe = as.character(entrez_universe$entrezgene_id))
```
```{r}
#BiocManager::install("org.Mm.eg.db")

gseKEGG_13 <- setReadable(gseKEGG_13, 'org.Mm.eg.db', 'ENTREZID')
gseKEGG_13_dotplot <- dotplot(gseKEGG_13)
gseKEGG_13_cnetplot <- cnetplot(gseKEGG_13)
gseKEGG_13_plot <- cowplot::plot_grid(gseKEGG_13_dotplot, gseKEGG_13_cnetplot)
gseKEGG_13_plot$data$geneID
gseKEGG_13_plot$data$Description
ggsave(plot = gseKEGG_13_plot, '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/cluster13_kegg.png', height=4, width=10)
write.csv(data.frame(gseKEGG_13), '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/cluster13_kegg.csv')

```

```{r}
gseKEGG_22 <- setReadable(gseKEGG_22, 'org.Mm.eg.db', 'ENTREZID')
gseKEGG_22_dotplot <- dotplot(gseKEGG_22, showCategory = 7)
gseKEGG_22_cnetplot <- cnetplot(gseKEGG_22)
gseKEGG_22_plot <- cowplot::plot_grid(gseKEGG_22_dotplot, gseKEGG_22_cnetplot, ncol = 1)
gseKEGG_22_plot$data$geneID
gseKEGG_22_plot$data$Description
ggsave(plot = gseKEGG_22_plot, '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/cluster22_kegg.png', height=15, width=15)
write.csv(data.frame(gseKEGG_22), '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/cluster22_kegg.csv')
```

```{r}
gseKEGG_27 <- setReadable(gseKEGG_27, 'org.Mm.eg.db', 'ENTREZID')
gseKEGG_27_dotplot <- dotplot(gseKEGG_27)
gseKEGG_27_cnetplot <- cnetplot(gseKEGG_27)
gseKEGG_27_plot <- cowplot::plot_grid(gseKEGG_27_dotplot, gseKEGG_27_cnetplot)
gseKEGG_27_plot$data$geneID
gseKEGG_27_plot$data$Description
ggsave(plot = gseKEGG_27_plot, '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/cluster27_kegg.png', height=4, width=15)
write.csv(data.frame(gseKEGG_27), '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/cluster27_kegg.csv')
```
BiocManager::install("pathview")
library('pathview')
```{r}
covid_genes <- data.frame(gseKEGG_22) %>% dplyr::filter(ID == 'mmu05171') 
covid_genes <- covid_genes$geneID
covid_genes <- unlist(strsplit(covid_genes, '\\/'))
covid_genes <- data.frame(covid_genes)
colnames(covid_genes) <- 'mgi_symbol'
covid_ids <- left_join(covid_genes, entrez_universe)
setwd('/Users/jwalla12/Repositories/dennery_scrna/scRNAseq')
pathview(gene.data = covid_ids$mgi_symbol,
         pathway.id = "mmu05171",
         species = "mmu",
         gene.idtype = "SYMBOL",
         gene.annotpkg = "org.Mm.eg.db"
         )

```







```{r}
#Idents(all_data.combined) <- all_data.combined$seurat_clusters
#expression_mat <- AverageExpression(all_data.combined, assays = 'SCT', add.ident = 'orig.ident')
```



```{r Make some specific comparisons...}

#Idents(all_data.combined) <- all_data.combined$orig.ident
#all_data.combined.AirD60_vs_O2D60 <- FindMarkers(all_data.combined, assay = 'SCT', ident.1 = 'AirD60', ident.2 = 'O2D60')





```


```{r Look at specific pathways of interest...}
#filter to get the genes in each of these kegg pathways or descrption keywords
glycolysis <- all_genes_bm %>% 
  filter(str_detect(kegg_enzyme, "^00010+"))

tca <- all_genes_bm %>% 
  filter(str_detect(kegg_enzyme, "^00020+"))

hmox <- all_genes_bm %>%
  filter(str_detect(mgi_description, "heme oxygenase"))

#make a dotplot of the hmox genes for each cluster:
hmox_dotplots <- lapply(seq_along(seurat.list), function(i) DotPlot(seurat.list[[i]], assay = 'SCT', features = c(hmox$mgi_symbol)) + RotatedAxis() + ggtitle(paste0("Heme oxygenase genes - Seurat Cluster ", names(seurat.list[i]))))

glycolysis_dotplots <- lapply(seq_along(seurat.list), function(i) DotPlot(seurat.list[[i]], assay = 'SCT', features = c(glycolysis$mgi_symbol)) + RotatedAxis() + ggtitle(paste0("Glycolysis genes - Seurat Cluster ", names(seurat.list[i]))))

tca_dotplots <- lapply(seq_along(seurat.list), function(i) DotPlot(seurat.list[[i]], assay = 'SCT', features = c(unique(tca$mgi_symbol))) + RotatedAxis() + ggtitle(paste0("TCA genes - Seurat Cluster ", names(seurat.list[i]))))

hmox_dotplots
glycolysis_dotplots
tca_dotplots
#save the dotplots
#lapply(seq_along(hmox_dotplots), function(i) ggsave(plot = hmox_dotplots[[i]], file = paste0("/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/figures/hmox_dotplots/cluster_", i-1, "_hmox_dotplot.pdf"), width = 11,  height = 7))

#save the dotplots
#lapply(seq_along(glycolysis_dotplots), function(i) ggsave(plot = glycolysis_dotplots[[i]], file = paste0("/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/figures/glycolysis_dotplots/cluster_", i-1, "_glycolysis_dotplot.pdf"), width = 15,  height = 7))

#save the dotplots
#lapply(seq_along(tca_dotplots), function(i) ggsave(plot = tca_dotplots[[i]], file = paste0("/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/figures/tca_dotplots/cluster_", i-1, "_tca_dotplot.pdf"), width = 11,  height = 7))
```









```{r}
#AT1_allmarkers <- FindAllMarkers(all_data.combined_AT1)
#write.table(AT1_allmarkers, '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/AT1_allmarkers.txt')

counts <- data.frame(all_data.combined@assays$SCT@counts) #these are the SCTransform corrected counts
counts$gene <- rownames(counts)
counts_melt <- melt(counts, id.vars = 'gene')
counts_melt <- tidyr::separate(data= counts_melt, col = variable, into = c('orig.ident', 'umi'), sep = "_", remove = F) %>% tidyr::separate(col = orig.ident, into = c('treat', 'time'), sep = "D", remove = F)

counts_melt$treat <- as.factor(counts_melt$treat) %>% relevel(ref = 'Air')
counts_melt$time <- as.factor(counts_melt$time) %>% relevel(ref = '7')

counts_melt <- rename(counts_melt, 'value' = 'counts')

# Build the linear model
model  <- lm(counts ~ treat*time,
             data = data.frame(counts_melt))

# the residuals are not normally distributed (p-value is significant)
counts_melt %>%
  group_by(treat, time) %>%
  slice_sample(n = 5000) %>%
  shapiro_test(counts)

counts_melt %>%
  slice_sample(n = 5000) %>%
  levene_test(counts ~ treat*time)

counts_melt_aov <- counts_melt %>%
  slice_sample(n = 5000) %>%
  anova_test(counts ~ treat*time)

counts_melt_aov

counts_melt %>%
  group_by(treat)%>%
  anova_test(counts ~ time, error = model)
```


```{r}
#AT1_allmarkers <- FindAllMarkers(all_data.combined_AT1)
#write.table(AT1_allmarkers, '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/AT1_allmarkers.txt')

AT1_counts <- data.frame(all_data.combined_AT1@assays$SCT@counts) #these are the SCTransform corrected counts
AT1_counts$gene <- rownames(AT1_counts)
AT1_counts_melt <- melt(AT1_counts, id.vars = 'gene')
AT1_counts_melt <- tidyr::separate(data= AT1_counts_melt, col = variable, into = c('orig.ident', 'umi'), sep = "_", remove = F) %>% tidyr::separate(col = orig.ident, into = c('treat', 'time'), sep = "D", remove = F)

AT1_counts_melt$treat <- as.factor(AT1_counts_melt$treat) %>% relevel(ref = 'Air')
AT1_counts_melt$time <- as.factor(AT1_counts_melt$time) %>% relevel(ref = '7')

AT1_counts_melt <- rename(AT1_counts_melt, 'value' = 'counts')

# Build the linear model
AT1_model  <- lm(counts ~ treat*time,
             data = data.frame(AT1_counts_melt))

# the residuals are not normally distributed (p-value is significant)
AT1_counts_melt %>%
  group_by(treat, time) %>%
  slice_sample(n = 5000) %>%
  shapiro_test(counts)

AT1_counts_melt %>%
  slice_sample(n = 5000) %>%
  levene_test(counts ~ treat*time)

AT1_counts_melt_aov <- AT1_counts_melt %>%
  slice_sample(n = 5000) %>%
  anova_test(counts ~ treat*time)

AT1_counts_melt_aov

AT1_counts_melt %>%
  group_by(treat)%>%
  anova_test(counts ~ time, error = AT1_model)
```
# So, there are definitely some interactions between time and treatment.


#DefaultAssay(all_data.combined_AT1) <- 'SCT'
```{r}
#O2D7_v_AirD7 <- FindMarkers(all_data.combined, ident.1 = "O2D7", ident.2 = "AirD7", group.by = 'orig.ident', assay = 'SCT')
```


```{r}
#O2D7_v_AirD7_AT1 <- FindMarkers(all_data.combined, ident.1 = "O2D7-25", ident.2 = "AirD7-25", group.by = 'ident_clust')
```

```{r}
#AllMarkers <- FindAllMarkers(all_data.combined, group.by = 'orig.ident', test.use = 'bimod')

```

```{r}
#save.image(file = "/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/dennery.RData")
```

```{r}
#load(file = "/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/dennery.RData")
```

```{r}
#library(monocle)

```

```{r}
#DefaultAssay(all_data.combined) <- 'RNA'

```

```{r}

##all_data.combined_gene_metadata <- data.frame(rownames(all_data.combined@assays$RNA))
#colnames(all_data.combined_gene_metadata) <- 'gene_short_name'
#row.names(all_data.combined_gene_metadata) <- all_data.combined_gene_metadata$gene_short_name
```


```{r}
#ggsave(file = '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/figures_and_tables/figures/qc_figures/supplemental_figure_1_filtering_metrics.png', plot = wrapped_filtering)
```


#We identified a total of  45 clusters of lung cells along with corresponding marker genes (Fig 2Ax, Table 1x). 

#This included 19 cell types  that included epithelial cells, endothelial cells, mesenchymal cells, and immune cells (Fig 2B and 2Cx, Table 2x)

