---
title: "Dennery lab scRNAseq"
author: "Joselynn"
date: "7/14/2020"
output: html_document
---


  
```{r Setup}
#install.packages('rstatix')
library(Seurat)
library(patchwork)
library(ggplot2)
library(dplyr)
library(future)
library(rsvd)
library(reshape2)
library(monocle3)
library(rstatix)
library(ggpubr)
library(biomaRt)
library(stringr)
set.seed(61)
options(future.globals.maxSize = 4000 * 1024^2)
setwd('/Users/jwalla12/Repositories/dennery_scrna/scRNAseq')
#source('/Users/jwalla12/Repositories/FIt-SNE/fast_tsne.R', chdir=T)
```

```{r Load Data}
# Read 10X
AirD7.data <- Read10X(data.dir ="/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/AirD7out/filtered_feature_bc_matrix/")
AirD60.data <- Read10X(data.dir ="/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/AirD60out/filtered_feature_bc_matrix/")
O2D7.data <- Read10X(data.dir ="/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/O2D7out/filtered_feature_bc_matrix/")
O2D60.data <- Read10X(data.dir ="/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/O2D60out/filtered_feature_bc_matrix/")
#SD7.data <- Read10X(data.dir ="./SD7out/filtered_feature_bc_matrix/")

# Create Seurat objects
#Include features (e.g. genes) detected in at least 3 cells and cells with at least 200 features.

AirD7.data <- CreateSeuratObject(counts = AirD7.data,  project = "AirD7",min.cells = 3, min.features = 200)
O2D7.data <- CreateSeuratObject(counts = O2D7.data,  project = "O2D7",min.cells = 3, min.features = 200)
AirD60.data <- CreateSeuratObject(counts = AirD60.data,  project = "AirD60",min.cells = 3, min.features = 200)
O2D60.data <- CreateSeuratObject(counts = O2D60.data,  project = "O2D60",min.cells = 3, min.features = 200)
#SD7.data <- CreateSeuratObject(counts = SD7.data,  project = "SD7",min.cells = 3, min.features = 200)

# Merge the data
#all_data <- merge(x = AirD7.data, y = c(O2D7.data, SD7.data, AirD60.data, O2D60.data), add.cell.ids = c("AirD7", "O2D7", "SD7", "AirD60", "O2D60"), project = 'dennery')
all_data <- merge(x = AirD7.data, y = c(O2D7.data, AirD60.data, O2D60.data), add.cell.ids = c("AirD7", "O2D7", "AirD60", "O2D60"), project = 'dennery')

```

```{r QC Plots}
# Calculate the percentage of counts that belong to mitochondrial genes
all_data[["percent.mt"]] <- PercentageFeatureSet(all_data, pattern = "^mt-")

#Set levels
Idents(all_data) <- factor(Idents(all_data), levels = c('AirD7', 'O2D7', 'AirD60', 'O2D60'))

# QC Plots
ggsave(file = '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/figures_and_tables/figures/qc_figures/nFeature_RNA_vlnplot.png', plot = VlnPlot(all_data, features = "nFeature_RNA"))
ggsave(file = '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/figures_and_tables/figures/qc_figures/nCount_RNA_vlnplot.png', plot = VlnPlot(all_data, features = "nCount_RNA"))
ggsave(file = '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/figures_and_tables/figures/qc_figures/percent.mt_vlnplot.png', plot = VlnPlot(all_data, features = "percent.mt"))

FeatureScatter(all_data, feature1 = "nCount_RNA", feature2 = "percent.mt")
FeatureScatter(all_data, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

# one QC metric to look at is the number of unique genes detected per cell.
# low-quality/empty droplets will have very few genes or features.
# cell doublets or multiplets will have very high count.
# it looks like there's a cluster of cells in AirD7 and O2D7 where the features and counts are both low -- e.g., there's cells where there's only a few lowly-expressed genes.
# 8000 seems like a reasonable upper bound on the features.
#https://www.bioinformatics.babraham.ac.uk/training/10XRNASeq/seurat_workflow.html
qc.metrics <- dplyr::as_tibble(all_data[[c("nCount_RNA","nFeature_RNA","percent.mt", "orig.ident")]],rownames="Cell.Barcode") 
qc.metrics

mplot <- qc.metrics %>%
  arrange(percent.mt) %>%
  ggplot(aes(nCount_RNA,nFeature_RNA,colour=percent.mt)) + 
  geom_point() + 
  scale_color_gradientn(colors=c("black","blue","green2","red","yellow")) +
  ggtitle("QC metrics") +
  geom_hline(yintercept = 700) +
  geom_hline(yintercept = 8000) 

ident_plot <- qc.metrics %>%
  arrange(percent.mt) %>%
  ggplot(aes(nCount_RNA,nFeature_RNA,color=factor(orig.ident))) + 
  geom_point(aes(alpha = 0.5)) + 
  ggtitle("QC metrics") +
  scale_color_manual(values=c("#E69F00", "#56B4E9", "#009E73", "#F0E442")) +
  geom_hline(yintercept = 700) +
  geom_hline(yintercept = 8000) 

wrap_plots(mplot, ident_plot, ncol = 2, guides = "collect")
```

```{r Filtering, normalization}

# Using a 5% mtDNA cutoff... as per https://www.biorxiv.org/content/10.1101/2020.02.20.958793v1.full.pdf
all_data_sub <- subset(all_data, subset = nFeature_RNA > 700 & nFeature_RNA < 8000 & percent.mt < 5)

all_data #20755 features across 16013 samples within 1 assay
all_data_sub #20755 features across 13199 samples within 1 assay 

# Post-filtering  QC plots

FeatureScatter(all_data_sub, feature1 = "nCount_RNA", feature2 = "percent.mt")
FeatureScatter(all_data_sub, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

# SCTransform to normalize, integrate, make some plots
#https://github.com/ChristophH/sctransform/issues/25 for explanation of why I'm not worried about 'iteration limit reached' warning messages.
all_data.list <- SplitObject(all_data_sub, split.by="orig.ident")

#sctransform
all_data.list <- lapply(all_data.list, function(x) {x <- SCTransform(x,verbose=FALSE)})
                        
#all the integration steps
all_data.features <- SelectIntegrationFeatures(object.list=all_data.list,nfeatures=3000)
all_data.list <- PrepSCTIntegration(object.list=all_data.list, anchor.features=all_data.features, verbose=FALSE)
all_data.int.anchors <- FindIntegrationAnchors(all_data.list, normalization.method="SCT", anchor.features=all_data.features, verbose=FALSE)
all_data.combined <- IntegrateData(anchorset=all_data.int.anchors, normalization.method="SCT", verbose=FALSE)

DefaultAssay(all_data.combined) <- "integrated"

all_data.combined <- RunPCA(all_data.combined, features = VariableFeatures(object = all_data.combined))
all_data.combined <- FindNeighbors(all_data.combined, dims = 1:50)
all_data.combined <- FindClusters(all_data.combined, resolution = 2)
all_data.combined <- RunUMAP(all_data.combined, dims=1:50)
all_data.combined <- RunTSNE(all_data.combined, dims=1:50, nthreads=4, max_iter=2000, seed.use=35) 
#DimPlot(all_data.combined, reduction="tsne")
#all_data.combined <- RunTSNE(all_data.combined, dims=1:50, tsne.method = "FIt-SNE", nthreads=4, max_iter=2000, seed.use=35)  #Could not find the following programs: fast_tsne

```

```{r Mouse cell atlas data}
# MCA
#Following https://satijalab.org/seurat/v3.1/mca.html vignette...

mca.matrix <- readRDS(file = "./MCA/MCA_merged_mat.rds")
mca.metadata <- read.csv("./MCA/MCA_All-batch-removed-assignments.csv")
mca.annotations <- read.csv("./MCA/MCA_CellAssignments.csv",row.names = 1)

mca.annotations.lung <- mca.annotations %>% dplyr::filter(Tissue == 'Lung')
mca.metadata.lung <- mca.metadata %>% dplyr::filter(Tissue == 'Lung')

rownames(mca.annotations.lung) <- mca.annotations.lung$Cell.name
rownames(mca.metadata.lung) <- mca.metadata.lung$Cell.name

mca.metadata.annotations <- dplyr::inner_join(mca.metadata.lung, mca.annotations.lung, by = "Cell.name")
mca.metadata.annotations <- mca.metadata.annotations %>% dplyr::select(Cell.name, ClusterID.x, Tissue.x, Batch.x, Cell.Barcode.x, Annotation)
colnames(mca.metadata.annotations) <- c('Cell.name', 'ClusterID', 'Tissue', 'Batch', 'Cell.Barcode', 'Annotation')
rownames(mca.metadata.annotations) <- mca.metadata.annotations$Cell.name

mca.metadata.annotations <- mca.metadata.annotations %>% dplyr::select(-Cell.name)

mca.matrix.lung <- mca.matrix[,grepl("^Lung_", colnames(mca.matrix))]

mca <- CreateSeuratObject(counts = mca.matrix.lung, meta.data = mca.metadata.annotations, project = "MouseCellAtlas")
# Only keep annotated cells
mca <- SCTransform(mca)
# Leaves us with 242k cells

all_data.list$mca <- mca
atlas.features <- SelectIntegrationFeatures(all_data.list) #might need to tweak this 
atlas.anchors <- FindTransferAnchors(reference=mca, query=all_data.combined, normalization.method="SCT", npcs=82, dims=1:50, features=atlas.features, verbose=FALSE)

atlas.predictions <- TransferData(anchorset=atlas.anchors, refdata=mca$Annotation, dims=1:50)
all_data.combined <- AddMetaData(all_data.combined, metadata=atlas.predictions)
all_data.combined$atlas_predicted_clusters <- sapply(sapply(all_data.combined$predicted.id, function(x) strsplit(x, " - ")), `[`, 1)

#saveRDS(all_data.combined, file=file.path("./all_data_combined.rds"))
#saveRDS(mca, file=file.path("./mca_sctransform.rds"))


#p1 #might need to make the plot window larger to see this properly...
```


```{r}
all_data.combined$orig.ident <- factor( x = all_data.combined$orig.ident, levels = c("AirD7", "O2D7", "AirD60", "O2D60"))
```


```{r}
p1 <- DimPlot(all_data.combined, reduction="tsne", group.by = "atlas_predicted_clusters") + NoLegend()
p1 <- LabelClusters(p1, id="atlas_predicted_clusters", repel=TRUE)

p2 <- DimPlot(all_data.combined, reduction="tsne", group.by = "seurat_clusters") + NoLegend()
p2 <- LabelClusters(p2, id="seurat_clusters", repel=TRUE)

wrap_plots(list(p1,p2), ncol = 2)

p3 <- DimPlot(all_data.combined, reduction="tsne", group.by = "atlas_predicted_clusters", split.by = "orig.ident") 
#p3 <- LabelClusters(p3, id="atlas_predicted_clusters", repel=TRUE)

p4 <- DimPlot(all_data.combined, reduction="tsne", group.by = "seurat_clusters", split.by = "orig.ident") 
#p4 <- LabelClusters(p4, id="seurat_clusters", repel=TRUE)

p3

p4
```


```{r}
all_data.combined <- readRDS("/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/all_data.combined.RDS")

# get a list of all the genes in the Seurat object.
all_genes <- data.frame(rownames(all_data.combined)) %>% rename(rownames.all_data.combined. = "mgi_symbol")

mouse_biomart <- useMart("ensembl", dataset = "mmusculus_gene_ensembl")

mouse_biomart_data <- getBM(mart = mouse_biomart, attributes = c("mgi_description", "mgi_symbol", "kegg_enzyme"))

all_genes_bm <- right_join(mouse_biomart_data, all_genes)
```

```{r How well do the Seurat clusters and the predicted clusters correspond?}
concordance <- data.frame(x=all_data.combined$atlas_predicted_clusters, y=all_data.combined$seurat_clusters)

normalize <- function(x) {
return ((x - min(x)) / (max(x) - min(x)))
}

conc_plot_data <- melt(table(concordance)) %>% group_by(y) %>% mutate(value=normalize(value))

conc_plot_data$y <- as.factor(conc_plot_data$y)

conc_plot <- ggplot(conc_plot_data, aes(x, y, fill = value)) +
  geom_tile(colour = 'black') + 
  theme(panel.grid = element_blank(),
        axis.title.y = element_blank(),
        axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank()) +
  scale_fill_gradientn(
    colors = c("white", "darkred"),
    limits = range(conc_plot_data$value)) +
  theme(panel.grid = element_blank(),
        axis.text.x=element_text(angle=90),
        axis.ticks = element_blank())
```


```{r Look at marker genes across clusters e.g., which genes are most important for distinguishing the cell clusters -- we were using the integrated data previously, but I think using the 'SCT' normalized data is a better idea.}
Idents(all_data.combined) <- all_data.combined$seurat_clusters

# find markers for every cluster compared to all remaining cells, report only the positive ones
all_data.combined.markers.sct <- FindAllMarkers(all_data.combined, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25, assay = 'SCT')
all_data.combined.markers.integrated <- FindAllMarkers(all_data.combined, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25, assay = 'integrated')

# Add the gene information to each cluster..
all_data.combined.markers.sct <- left_join(all_data.combined.markers.sct, all_genes_bm, by = c('gene' = 'mgi_symbol'))
all_data.combined.markers.integrated <- left_join(all_data.combined.markers.integrated, all_genes_bm, by = c('gene' = 'mgi_symbol'))

#split up the markers by cluster..
cluster_levels <- levels(all_data.combined.markers.sct$cluster)
all_data.combined.markers.sct.split <- lapply(seq_along(cluster_levels), function(i) all_data.combined.markers.sct %>% dplyr::filter(cluster == cluster_levels[i]))
all_data.combined.markers.integrated.split <- lapply(seq_along(cluster_levels), function(i) all_data.combined.markers.integrated %>% dplyr::filter(cluster == cluster_levels[i]))

#doing some filtering to look at just the 'top markers', but probably want to only use these for making figures.
top5_markers <- all_data.combined.markers.sct %>% group_by(cluster) %>% top_n(n = 5, wt = avg_logFC)

top5_markers$rank <- rep(1:5, length.out = nrow(top5_markers))

clust_plot <- ggplot(top5_markers, aes(x = rank, y = cluster, fill = avg_logFC)) + 
  geom_tile() +
  theme(panel.grid = element_blank(),
        axis.title.y = element_blank(),
        axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank()) +
  geom_text(aes(label = gene, size = 8)) +
  xlab("") +
  scale_fill_gradientn(
    colors = c("white", "purple3"),
    limits = range(top5_markers$avg_logFC)) +
  theme_minimal()



```


# AT1, AT2, and B cells are in similar clusters (does that make sense? are they similar cell types?)
# AT1 cells are in cluster 25, top markers: Ndst1, Mal2, Tead1, Agrn, Dag1, between 35 and 52 cells in each experiment
# AT2 cells are in a few clusters -- 3, 11, 13, 22, and 27. 22 and 27 have other cell types too (B cells, AT1 cells, alveolar macrophages), but are mainly AT2.
# Cluster 3 markers: Cat, Irx3, Pla2g1b, Ctsh, Elf5
# Cluster 11 markers: Kcnc3, LOC115487417, Ppp1r9a, Gm33122, St6galnac2
# Cluster 13 markers: Fasn, Tgoln1, Il33, Lrg2, Lrp2
# Cluster 22 markers: 



# how many cells are in each cell type?
```{r}
DefaultAssay(all_data.combined) <- 'SCT'

all_data.combined_meta.data <- tibble(all_data.combined@meta.data) 

counts_per_cluster_and_ident <- all_data.combined_meta.data %>% group_by(seurat_clusters, orig.ident) %>% tally() 

counts_per_cluster_and_ident_spread <- counts_per_cluster_and_ident %>% tidyr::spread(key=orig.ident, value=n)

nrow(all_data.combined_meta.data)
```

```{r}
table(all_data.combined@meta.data$seurat_clusters, all_data.combined@meta.data$orig.ident)

table(all_data.combined@active.ident, all_data.combined@meta.data$orig.ident)
```

#https://github.com/satijalab/seurat/issues/738
```{r}
count_plot <- ggplot(counts_per_cluster_and_ident, aes(x = orig.ident, y = seurat_clusters, fill = n)) + 
  geom_tile() +
  theme(panel.grid = element_blank(),
        axis.title.y = element_blank(),
        axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank()) +
  geom_text(aes(label = n)) +
  xlab("") +
  scale_fill_gradientn(
    colors = c("white", "grey"),
    limits = range(counts_per_cluster_and_ident$n)) +
  theme_minimal()
```




```{r}
sessionInfo()
```

```{r}
#saveRDS(all_data.combined, file = "/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/all_data.combined.RDS")
```

```{r}
#write.table((table(all_data.combined@meta.data$seurat_clusters, all_data.combined@meta.data$orig.ident)), "/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/cells_per_seurat_cluster.txt")
```

```{r}
#write.table((table(all_data.combined@meta.data$predicted.id, all_data.combined@meta.data$orig.ident)), "/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/cells_per_mca_cluster_annot.txt")
```







```{r Differential Expression}
# It seems like folks are interested in a few clusters:
# Cluster 25 - AT1 cells
# Clusters 3, 11, 13, 22, and 27 are the AT2 cell types (22 and 27 have other cell types too).
# Ciliated cells are in Cluster 28
# Club cells are in Cluster 39

levels(all_data.combined$orig.ident)

#make a new column in all_data.combined@meta.data that combines the seurat cluster and orig.ident
all_data.combined@meta.data$ident_clust <- paste0(all_data.combined@meta.data$orig.ident, "-", all_data.combined@meta.data$seurat_clusters)

all_data.combined@meta.data$treat <- tidyr::separate(data = all_data.combined@meta.data, 
                col = orig.ident,
                into = c('treat','time'),
                sep = 'D',
                remove = F)$treat


all_data.combined@meta.data$time <- tidyr::separate(data = all_data.combined@meta.data, 
                col = orig.ident,
                into = c('treat','time'),
                sep = 'D',
                remove = F)$time
```

```{r}
Idents(all_data.combined) <- 'orig.ident'
DefaultAssay(all_data.combined) <- 'SCT'
```



# This chunk will take the Seurat object as an input and filter is such that each seurat cluster is in its own seurat object and then findallmarkers, then make dotplots of the top markers across the experimental conditions
```{r}
#set the levels
clusters <- levels(all_data.combined@meta.data$seurat_clusters)

construct <- unlist(lapply(seq_along(clusters), function(i) paste0("all_data.combined_", clusters[i], "<- subset(all_data.combined, subset = seurat_clusters =='", clusters[i], "')")))

seurat.list <- lapply(seq_along(construct), function(i) eval(parse(text = construct[i])))

names(seurat.list) <- c(levels(all_data.combined@meta.data$seurat_clusters))

seurat.list.markers <- lapply(seurat.list, FindAllMarkers, min.pct = 0.25, logfc.threshold = 0.25, assay = 'SCT')

#Add the gene explanations to each file..
seurat.list.markers.ids <- lapply(seq_along(seurat.list.markers), function(i) left_join(seurat.list.markers[[i]], all_genes_bm, by = c('gene' = 'mgi_symbol')))

#write the findallmarkers outputs for each cluster to a file
lapply(seq_along(seurat.list.markers.ids), function(i) write.csv(seurat.list.markers.ids[[i]], file = paste0("/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/figures_and_tables/tables/condition_markers/cluster_", i-1, "_per_condition_markers.csv"), row.names = F))

get_top_markers <- function(seurat_markers){
  output <- seurat_markers %>% group_by(cluster)  %>% top_n(n = 10, wt = avg_logFC)
  return(output)
}

seurat.list.top.markers <- lapply(seurat.list.markers, get_top_markers)

seurat_list_dotplots <- lapply(seq_along(seurat.list), function(i) DotPlot(seurat.list[[i]], features = c(unique(seurat.list.top.markers[[i]]$gene))) + RotatedAxis() + ggtitle(paste0("Seurat Cluster ", names(seurat.list[i]))))

#lapply(seq_along(seurat_list_dotplots), function(i) ggsave(plot = seurat_list_dotplots[[i]], file = paste0("/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/figures/cluster_", i-1, "_dotplot.pdf"), width = 11,  height = 7))

```


```{r Look at specific pathways of interest...}
#filter to get the genes in each of these kegg pathways or descrption keywords
glycolysis <- all_genes_bm %>% 
  filter(str_detect(kegg_enzyme, "^00010+"))

tca <- all_genes_bm %>% 
  filter(str_detect(kegg_enzyme, "^00020+"))

hmox <- all_genes_bm %>%
  filter(str_detect(mgi_description, "heme oxygenase"))

#make a dotplot of the hmox genes for each cluster:
hmox_dotplots <- lapply(seq_along(seurat.list), function(i) DotPlot(seurat.list[[i]], assay = 'SCT', features = c(hmox$mgi_symbol)) + RotatedAxis() + ggtitle(paste0("Heme oxygenase genes - Seurat Cluster ", names(seurat.list[i]))))

glycolysis_dotplots <- lapply(seq_along(seurat.list), function(i) DotPlot(seurat.list[[i]], assay = 'SCT', features = c(glycolysis$mgi_symbol)) + RotatedAxis() + ggtitle(paste0("Glycolysis genes - Seurat Cluster ", names(seurat.list[i]))))

tca_dotplots <- lapply(seq_along(seurat.list), function(i) DotPlot(seurat.list[[i]], assay = 'SCT', features = c(unique(tca$mgi_symbol))) + RotatedAxis() + ggtitle(paste0("TCA genes - Seurat Cluster ", names(seurat.list[i]))))

#save the dotplots
lapply(seq_along(hmox_dotplots), function(i) ggsave(plot = hmox_dotplots[[i]], file = paste0("/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/figures/hmox_dotplots/cluster_", i-1, "_hmox_dotplot.pdf"), width = 11,  height = 7))

#save the dotplots
lapply(seq_along(glycolysis_dotplots), function(i) ggsave(plot = glycolysis_dotplots[[i]], file = paste0("/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/figures/glycolysis_dotplots/cluster_", i-1, "_glycolysis_dotplot.pdf"), width = 15,  height = 7))

#save the dotplots
lapply(seq_along(tca_dotplots), function(i) ggsave(plot = tca_dotplots[[i]], file = paste0("/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/figures/tca_dotplots/cluster_", i-1, "_tca_dotplot.pdf"), width = 11,  height = 7))
```


```{r Make some specific comparisons...}
AT1_D60_Air_v_O2.markers <- FindMarkers(all_data.combined_AT1, min.pct = 0.5, logfc.threshold = 0.25, assay = 'SCT', ident.1 = 'O2D60', ident.2 = 'AirD60')

DotPlot(all_data.combined_AT1, features = unique(rownames(AT1_D60_Air_v_O2.markers)))  + RotatedAxis()

```






```{r}
#AT1_allmarkers <- FindAllMarkers(all_data.combined_AT1)
#write.table(AT1_allmarkers, '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/AT1_allmarkers.txt')

counts <- data.frame(all_data.combined@assays$SCT@counts) #these are the SCTransform corrected counts
counts$gene <- rownames(counts)
counts_melt <- melt(counts, id.vars = 'gene')
counts_melt <- tidyr::separate(data= counts_melt, col = variable, into = c('orig.ident', 'umi'), sep = "_", remove = F) %>% tidyr::separate(col = orig.ident, into = c('treat', 'time'), sep = "D", remove = F)

counts_melt$treat <- as.factor(counts_melt$treat) %>% relevel(ref = 'Air')
counts_melt$time <- as.factor(counts_melt$time) %>% relevel(ref = '7')

counts_melt <- rename(counts_melt, 'value' = 'counts')

# Build the linear model
model  <- lm(counts ~ treat*time,
             data = data.frame(counts_melt))

# the residuals are not normally distributed (p-value is significant)
counts_melt %>%
  group_by(treat, time) %>%
  slice_sample(n = 5000) %>%
  shapiro_test(counts)

counts_melt %>%
  slice_sample(n = 5000) %>%
  levene_test(counts ~ treat*time)

counts_melt_aov <- counts_melt %>%
  slice_sample(n = 5000) %>%
  anova_test(counts ~ treat*time)

counts_melt_aov

counts_melt %>%
  group_by(treat)%>%
  anova_test(counts ~ time, error = model)
```


```{r}
#AT1_allmarkers <- FindAllMarkers(all_data.combined_AT1)
#write.table(AT1_allmarkers, '/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/AT1_allmarkers.txt')

AT1_counts <- data.frame(all_data.combined_AT1@assays$SCT@counts) #these are the SCTransform corrected counts
AT1_counts$gene <- rownames(AT1_counts)
AT1_counts_melt <- melt(AT1_counts, id.vars = 'gene')
AT1_counts_melt <- tidyr::separate(data= AT1_counts_melt, col = variable, into = c('orig.ident', 'umi'), sep = "_", remove = F) %>% tidyr::separate(col = orig.ident, into = c('treat', 'time'), sep = "D", remove = F)

AT1_counts_melt$treat <- as.factor(AT1_counts_melt$treat) %>% relevel(ref = 'Air')
AT1_counts_melt$time <- as.factor(AT1_counts_melt$time) %>% relevel(ref = '7')

AT1_counts_melt <- rename(AT1_counts_melt, 'value' = 'counts')

# Build the linear model
AT1_model  <- lm(counts ~ treat*time,
             data = data.frame(AT1_counts_melt))

# the residuals are not normally distributed (p-value is significant)
AT1_counts_melt %>%
  group_by(treat, time) %>%
  slice_sample(n = 5000) %>%
  shapiro_test(counts)

AT1_counts_melt %>%
  slice_sample(n = 5000) %>%
  levene_test(counts ~ treat*time)

AT1_counts_melt_aov <- AT1_counts_melt %>%
  slice_sample(n = 5000) %>%
  anova_test(counts ~ treat*time)

AT1_counts_melt_aov

AT1_counts_melt %>%
  group_by(treat)%>%
  anova_test(counts ~ time, error = AT1_model)
```
# So, there are definitely some interactions between time and treatment.


DefaultAssay(all_data.combined_AT1) <- 'SCT'
```{r}
O2D7_v_AirD7 <- FindMarkers(all_data.combined, ident.1 = "O2D7", ident.2 = "AirD7", group.by = 'orig.ident', assay = 'SCT')
```


```{r}
#O2D7_v_AirD7_AT1 <- FindMarkers(all_data.combined, ident.1 = "O2D7-25", ident.2 = "AirD7-25", group.by = 'ident_clust')
```

```{r}
#AllMarkers <- FindAllMarkers(all_data.combined, group.by = 'orig.ident', test.use = 'bimod')

```

```{r}
#save.image(file = "/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/dennery.RData")
```

```{r}
#load(file = "/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/dennery.RData")
```

```{r}
#library(monocle)

```

```{r}
DefaultAssay(all_data.combined) <- 'RNA'

```

```{r}

all_data.combined_gene_metadata <- data.frame(rownames(all_data.combined@assays$RNA))
colnames(all_data.combined_gene_metadata) <- 'gene_short_name'
row.names(all_data.combined_gene_metadata) <- all_data.combined_gene_metadata$gene_short_name
```



