---
title: "Dennery lab scRNAseq"
author: "Joselynn"
date: "7/14/2020"
output: html_document
---


  
```{r Setup}
library(Seurat)
library(patchwork)
library(ggplot2)
library(dplyr)
library(future)
library(rsvd)
library(reshape2)

set.seed(61)
options(future.globals.maxSize = 4000 * 1024^2)
setwd('/Users/jwalla12/Repositories/dennery_scrna/scRNAseq')
#source('/Users/jwalla12/Repositories/FIt-SNE/fast_tsne.R', chdir=T)
```

```{r}
# Load data}

# Read 10X
AirD7.data <- Read10X(data.dir ="./AirD7out/filtered_feature_bc_matrix/")
AirD60.data <- Read10X(data.dir ="./AirD60out/filtered_feature_bc_matrix/")
O2D7.data <- Read10X(data.dir ="./O2D7out/filtered_feature_bc_matrix/")
O2D60.data <- Read10X(data.dir ="./O2D60out/filtered_feature_bc_matrix/")
#SD7.data <- Read10X(data.dir ="./SD7out/filtered_feature_bc_matrix/")

# Create Seurat objects
#Include features (e.g. genes) detected in at least 3 cells and cells with at least 200 features.

AirD7.data <- CreateSeuratObject(counts = AirD7.data,  project = "AirD7",min.cells = 3, min.features = 200)
O2D7.data <- CreateSeuratObject(counts = O2D7.data,  project = "O2D7",min.cells = 3, min.features = 200)
AirD60.data <- CreateSeuratObject(counts = AirD60.data,  project = "AirD60",min.cells = 3, min.features = 200)
O2D60.data <- CreateSeuratObject(counts = O2D60.data,  project = "O2D60",min.cells = 3, min.features = 200)
#SD7.data <- CreateSeuratObject(counts = SD7.data,  project = "SD7",min.cells = 3, min.features = 200)

# Merge the data
#all_data <- merge(x = AirD7.data, y = c(O2D7.data, SD7.data, AirD60.data, O2D60.data), add.cell.ids = c("AirD7", "O2D7", "SD7", "AirD60", "O2D60"), project = 'dennery')
all_data <- merge(x = AirD7.data, y = c(O2D7.data, AirD60.data, O2D60.data), add.cell.ids = c("AirD7", "O2D7", "AirD60", "O2D60"), project = 'dennery')

```

```{r}
# Calculate the percentage of counts that belong to mitochondrial genes
all_data[["percent.mt"]] <- PercentageFeatureSet(all_data, pattern = "^mt-")

# QC Plots
VlnPlot(all_data, features = "nFeature_RNA")
VlnPlot(all_data, features = "nCount_RNA")
VlnPlot(all_data, features="percent.mt")

FeatureScatter(all_data, feature1 = "nCount_RNA", feature2 = "percent.mt")
FeatureScatter(all_data, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

# one QC metric to look at is the number of unique genes detected per cell.
# low-quality/empty droplets will have very few genes or features.
# cell doublets or multiplets will have very high count.
# it looks like there's a cluster of cells in AirD7 and O2D7 where the features and counts are both low -- e.g., there's cells where there's only a few lowly-expressed genes.
# 8000 seems like a reasonable upper bound on the features.
#https://www.bioinformatics.babraham.ac.uk/training/10XRNASeq/seurat_workflow.html
qc.metrics <- dplyr::as_tibble(all_data[[c("nCount_RNA","nFeature_RNA","percent.mt", "orig.ident")]],rownames="Cell.Barcode") 
qc.metrics
```


```{r}
mplot <- qc.metrics %>%
  arrange(percent.mt) %>%
  ggplot(aes(nCount_RNA,nFeature_RNA,colour=percent.mt)) + 
  geom_point() + 
  scale_color_gradientn(colors=c("black","blue","green2","red","yellow")) +
  ggtitle("QC metrics") +
  geom_hline(yintercept = 700) +
  geom_hline(yintercept = 8000) 

ident_plot <- qc.metrics %>%
  arrange(percent.mt) %>%
  ggplot(aes(nCount_RNA,nFeature_RNA,color=factor(orig.ident))) + 
  geom_point(aes(alpha = 0.5)) + 
  ggtitle("QC metrics") +
  scale_color_manual(values=c("#E69F00", "#56B4E9", "#009E73", "#F0E442")) +
  geom_hline(yintercept = 700) +
  geom_hline(yintercept = 8000) 
```

```{r}
wrap_plots(mplot, ident_plot, ncol = 2, guides = "collect")
```


```{r}
# Filtering data

# Using a 5% mtDNA cutoff... as per https://www.biorxiv.org/content/10.1101/2020.02.20.958793v1.full.pdf
all_data_sub <- subset(all_data, subset = nFeature_RNA > 700 & nFeature_RNA < 8000 & percent.mt < 5)

all_data #20755 features across 16013 samples within 1 assay
all_data_sub #20755 features across 13199 samples within 1 assay 

# Post-filtering  QC plots

FeatureScatter(all_data_sub, feature1 = "nCount_RNA", feature2 = "percent.mt")
FeatureScatter(all_data_sub, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

# SCTransform to normalize, integrate, make some plots
#https://github.com/ChristophH/sctransform/issues/25 for explanation of why I'm not worried about 'iteration limit reached' warning messages.
all_data.list <- SplitObject(all_data_sub, split.by="orig.ident")

#sctransform
all_data.list <- lapply(all_data.list, function(x) {x <- SCTransform(x,verbose=FALSE)})
                        
#all the integration steps
all_data.features <- SelectIntegrationFeatures(object.list=all_data.list,nfeatures=3000)
all_data.list <- PrepSCTIntegration(object.list=all_data.list, anchor.features=all_data.features, verbose=FALSE)
all_data.int.anchors <- FindIntegrationAnchors(all_data.list, normalization.method="SCT", anchor.features=all_data.features, verbose=FALSE)
all_data.combined <- IntegrateData(anchorset=all_data.int.anchors, normalization.method="SCT", verbose=FALSE)

DefaultAssay(all_data.combined) <- "integrated"

all_data.combined <- RunPCA(all_data.combined, features = VariableFeatures(object = all_data.combined))
all_data.combined <- FindNeighbors(all_data.combined, dims = 1:50)
all_data.combined <- FindClusters(all_data.combined, resolution = 2)
all_data.combined <- RunUMAP(all_data.combined, dims=1:50)
all_data.combined <- RunTSNE(all_data.combined, dims=1:50, nthreads=4, max_iter=2000, seed.use=35) 
#DimPlot(all_data.combined, reduction="tsne")
#all_data.combined <- RunTSNE(all_data.combined, dims=1:50, tsne.method = "FIt-SNE", nthreads=4, max_iter=2000, seed.use=35)  #Could not find the following programs: fast_tsne


# MCA
#Following https://satijalab.org/seurat/v3.1/mca.html vignette...

mca.matrix <- readRDS(file = "./MCA/MCA_merged_mat.rds")
mca.metadata <- read.csv("./MCA/MCA_All-batch-removed-assignments.csv")
mca.annotations <- read.csv("./MCA/MCA_CellAssignments.csv",row.names = 1)

mca.annotations.lung <- mca.annotations %>% dplyr::filter(Tissue == 'Lung')
mca.metadata.lung <- mca.metadata %>% dplyr::filter(Tissue == 'Lung')

rownames(mca.annotations.lung) <- mca.annotations.lung$Cell.name
rownames(mca.metadata.lung) <- mca.metadata.lung$Cell.name

mca.metadata.annotations <- dplyr::inner_join(mca.metadata.lung, mca.annotations.lung, by = "Cell.name")
mca.metadata.annotations <- mca.metadata.annotations %>% select(Cell.name, ClusterID.x, Tissue.x, Batch.x, Cell.Barcode.x, Annotation)
colnames(mca.metadata.annotations) <- c('Cell.name', 'ClusterID', 'Tissue', 'Batch', 'Cell.Barcode', 'Annotation')
rownames(mca.metadata.annotations) <- mca.metadata.annotations$Cell.name

mca.metadata.annotations <- mca.metadata.annotations %>% select(-Cell.name)

mca.matrix.lung <- mca.matrix[,grepl("^Lung_", colnames(mca.matrix))]

mca <- CreateSeuratObject(counts = mca.matrix.lung, meta.data = mca.metadata.annotations, project = "MouseCellAtlas")
# Only keep annotated cells
mca <- SCTransform(mca)
# Leaves us with 242k cells

all_data.list$mca <- mca
atlas.features <- SelectIntegrationFeatures(all_data.list) #might need to tweak this 
atlas.anchors <- FindTransferAnchors(reference=mca, query=all_data.combined, normalization.method="SCT", npcs=82, dims=1:50, features=atlas.features, verbose=FALSE)

atlas.predictions <- TransferData(anchorset=atlas.anchors, refdata=mca$Annotation, dims=1:50)
all_data.combined <- AddMetaData(all_data.combined, metadata=atlas.predictions)
all_data.combined$atlas_predicted_clusters <- sapply(sapply(all_data.combined$predicted.id, function(x) strsplit(x, " - ")), `[`, 1)

#saveRDS(all_data.combined, file=file.path("./all_data_combined.rds"))
#saveRDS(mca, file=file.path("./mca_sctransform.rds"))


#p1 #might need to make the plot window larger to see this properly...
```


```{r}
qc.metrics %>%
  arrange(percent.mt) %>%
  ggplot(aes(nCount_RNA,nFeature_RNA,color=factor(orig.ident))) + 
  geom_point() + 
  ggtitle("QC metrics") +
  scale_color_manual(values=c("#E69F00", "#56B4E9", "#009E73", "#F0E442")) +
  geom_hline(yintercept = 700) +
  geom_hline(yintercept = 8000) 
```


#tabula muris - skip the tabula annotations for now.
```{r}
#load("tabula_muris/R_objects/droplet_Lung_seurat_tiss.Robj", verbose=TRUE)

#tiss_update <- UpdateSeuratObject(tiss)

#tiss_sc <- SCTransform(tiss_update)

#all_data.list$tiss <- tiss_sc
#tabula.features <- SelectIntegrationFeatures(all_data.list) #might need to tweak this 
#tabula.anchors <- FindTransferAnchors(reference=tiss_sc, query=all_data.combined, normalization.method="SCT", npcs=82, dims=1:50, features=tabula.features, verbose=FALSE)

#tabula.predictions <- TransferData(anchorset=tabula.anchors, refdata=tiss_sc$cell_ontology_class, dims=1:50)
#all_data.combined <- AddMetaData(all_data.combined, metadata=tabula.predictions)
#all_data.combined$tabula_predicted_clusters <- sapply(sapply(all_data.combined$predicted.id, function(x) strsplit(x, " - ")), `[`, 1)

```

```{r}
all_data.combined$orig.ident <- factor( x = all_data.combined$orig.ident, levels = c("AirD7", "O2D7", "AirD60", "O2D60"))
```


```{r}
p1 <- DimPlot(all_data.combined, reduction="tsne", group.by = "atlas_predicted_clusters") + NoLegend()
p1 <- LabelClusters(p1, id="atlas_predicted_clusters", repel=TRUE)

p2 <- DimPlot(all_data.combined, reduction="tsne", group.by = "seurat_clusters") + NoLegend()
p2 <- LabelClusters(p2, id="seurat_clusters", repel=TRUE)

wrap_plots(list(p1,p2), ncol = 2)
```

```{r}
p3 <- DimPlot(all_data.combined, reduction="tsne", group.by = "atlas_predicted_clusters", split.by = "orig.ident") 
#p3 <- LabelClusters(p3, id="atlas_predicted_clusters", repel=TRUE)
```

```{r}
p4 <- DimPlot(all_data.combined, reduction="tsne", group.by = "seurat_clusters", split.by = "orig.ident") 
#p4 <- LabelClusters(p4, id="seurat_clusters", repel=TRUE)
```

```{r}
p3
```

```{r}
p4
```


```{r}
# find markers for every cluster compared to all remaining cells, report only the positive ones
all_data.combined.markers <- FindAllMarkers(all_data.combined, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

top5_markers <- all_data.combined.markers %>% group_by(cluster) %>% top_n(n = 5, wt = avg_logFC)

top5_markers$rank <- rep(1:5, times = 45)

clust_plot <- ggplot(top5_markers, aes(x = rank, y = cluster, fill = avg_logFC)) + 
  geom_tile() +
  theme(panel.grid = element_blank(),
        axis.title.y = element_blank(),
        axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank()) +
  geom_text(aes(label = gene, size = 8)) +
  xlab("") +
  scale_fill_gradientn(
    colors = c("white", "purple3"),
    limits = range(top5_markers$avg_logFC)) +
  theme_minimal()

concordance <- data.frame(x=all_data.combined$atlas_predicted_clusters, y=all_data.combined$seurat_clusters)

normalize <- function(x) {
return ((x - min(x)) / (max(x) - min(x)))
}

conc_plot_data <- melt(table(concordance)) %>% group_by(y) %>% mutate(value=normalize(value))

conc_plot_data$y <- as.factor(conc_plot_data$y)

conc_plot <- ggplot(conc_plot_data, aes(x, y, fill = value)) +
  geom_tile() + 
  theme(panel.grid = element_blank(),
        axis.title.y = element_blank(),
        axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank()) +
  scale_fill_gradientn(
    colors = c("white", "darkorange"),
    limits = range(conc_plot_data$value)) +
  theme(panel.grid = element_blank(),
        axis.text.x=element_text(angle=90),
        axis.ticks = element_blank())
```


```{r}
wrapped <- wrap_plots(clust_plot, conc_plot, ncol = 2, guides = "collect")
```

```{r}
table(all_data.combined@meta.data) %>% group_by(orig.ident) %>% table(seurat_clusters)
```


# AT1, AT2, and B cells are in similar clusters (does that make sense? are they similar cell types?)
# AT1 cells are in cluster 25, top markers: Ndst1, Mal2, Tead1, Agrn, Dag1, between 35 and 52 cells in each experiment
# AT2 cells are in a few clusters -- 3, 11, 13, 22, and 27. 22 and 27 have other cell types too (B cells, AT1 cells, alveolar macrophages), but are mainly AT2.
# Cluster 3 markers: Cat, Irx3, Pla2g1b, Ctsh, Elf5
# Cluster 11 markers: Kcnc3, LOC115487417, Ppp1r9a, Gm33122, St6galnac2
# Cluster 13 markers: Fasn, Tgoln1, Il33, Lrg2, Lrp2
# Cluster 22 markers: 



# how many cells are in each cell type?
```{r}
all_data.combined_meta.data <- tibble(all_data.combined@meta.data) 

counts_per_cluster_and_ident <- all_data.combined_meta.data %>% group_by(seurat_clusters, orig.ident) %>% tally() 
```

```{r}
table(all_data.combined@meta.data$seurat_clusters, all_data.combined@meta.data$orig.ident)

table(all_data.combined@active.ident, all_data.combined@meta.data$orig.ident)
```

#https://github.com/satijalab/seurat/issues/738
```{r}
count_plot <- ggplot(counts_per_cluster_and_ident, aes(x = orig.ident, y = seurat_clusters, fill = n)) + 
  geom_tile() +
  theme(panel.grid = element_blank(),
        axis.title.y = element_blank(),
        axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank()) +
  geom_text(aes(label = n)) +
  xlab("") +
  scale_fill_gradientn(
    colors = c("white", "grey"),
    limits = range(counts_per_cluster_and_ident$n)) +
  theme_minimal()
```

```{r}
more_plots <- wrap_plots(clust_plot, conc_plot, count_plot, ncol = 3, guides = "collect")
```


```{r}
counts_per_cluster_and_ident_spread <- counts_per_cluster_and_ident %>% tidyr::spread(key=orig.ident, value=n)
```


```{r}
sessionInfo()
```

```{r}
save.image(file = "/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/dennery.RData")
```

```{r}
saveRDS(all_data.combined, file = "/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/all_data.combined.RDS")
```

```{r}
write.table((table(all_data.combined@meta.data$seurat_clusters, all_data.combined@meta.data$orig.ident)), "/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/cells_per_seurat_cluster.txt")
```

```{r}
write.table((table(all_data.combined@meta.data$predicted.id, all_data.combined@meta.data$orig.ident)), "/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/cells_per_mca_cluster_annot.txt")
```

```{r}
load(file = "/Users/jwalla12/Repositories/dennery_scrna/scRNAseq/dennery.RData")
```


DotPlot(
  object,
  assay = NULL,
  features,
  cols = c("lightgrey", "blue"),
  col.min = -2.5,
  col.max = 2.5,
  dot.min = 0,
  dot.scale = 6,
  group.by = NULL,
  split.by = NULL,
  scale.by = "radius",
  scale.min = NA,
  scale.max = NA
)
```{r}
top_marker <- all_data.combined.markers %>% group_by(cluster) %>% top_n(n = 1, wt = avg_logFC)



DotPlot(all_data.combined, features = c(top_marker$gene))  + RotatedAxis()
```

```{r}

top5_markers[which(top5_markers$cluster == 1),]$gene

```
```{r}
grouped <- top5_markers %>% group_by(cluster) %>% select(gene) 
out <- split( df , f = df$userid )

out <- split( df , f = df$userid )
```
```{r}
clusters_list <- list(0:(length(unique(top5_markers$cluster))-1))
```

```{r}
markers_list <- lapply(seq_along(top5_markers$cluster), function(i) top5_markers[which(top5_markers$cluster == i),]$gene) 
```
lapply(seq_along(diatoms), function(i)